// ngx-auth-firebaseui/schematics/ng-add/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const tasks_1 = require("@angular-devkit/schematics/tasks");
const helpers_1 = require("../helpers");
const config_1 = require("@schematics/angular/utility/config");
const schematics_2 = require("@angular/cdk/schematics");
/** Loads the full version from the given Angular package gracefully. */
function loadPackageVersionGracefully() {
    try {
        console.log('ngx-auth-firebaseui version = ', require(`../package.json`).version);
        return require(`../package.json`).version;
    }
    catch (_a) {
        return null;
    }
}
// You don't have to export the function as default. You can also have more than one rule factory
// per file.
function addPackageJsonDependencies() {
    return (host, context) => {
        const ngCoreVersionTag = getPackageVersionFromPackageJson(host, '@angular/core');
        const dependencies = [
            { type: helpers_1.NodeDependencyType.Default, version: loadPackageVersionGracefully() || '4.0.3', name: 'ngx-auth-firebaseui' },
            { type: helpers_1.NodeDependencyType.Default, version: ngCoreVersionTag || '9.0.0', name: '@angular/animations' },
            { type: helpers_1.NodeDependencyType.Default, version: ngCoreVersionTag || '9.0.0', name: '@angular/forms' },
            { type: helpers_1.NodeDependencyType.Default, version: ngCoreVersionTag || '9.0.0', name: '@angular/router' },
            { type: helpers_1.NodeDependencyType.Default, version: '^9.0.0-beta.29', name: '@angular/flex-layout' },
            { type: helpers_1.NodeDependencyType.Default, version: '^5.4.2', name: '@angular/fire' },
            { type: helpers_1.NodeDependencyType.Default, version: '^7.8.1', name: 'firebase' }
        ];
        dependencies.forEach(dependency => {
            helpers_1.addPackageJsonDependency(host, dependency);
            context.logger.log('info', `✅️ Added "${dependency.name}" into ${dependency.type}`);
        });
        return host;
    };
}
exports.addPackageJsonDependencies = addPackageJsonDependencies;
function installPackageJsonDependencies() {
    return (host, context) => {
        context.addTask(new tasks_1.NodePackageInstallTask());
        context.logger.log('info', `🔍 Installing packages...`);
        return host;
    };
}
exports.installPackageJsonDependencies = installPackageJsonDependencies;
function addModuleToImports(options) {
    return (host, context) => {
        const workspace = config_1.getWorkspace(host);
        const project = schematics_2.getProjectFromWorkspace(workspace, options.project);
        // const x =
        //   `apiKey: 'your-firebase-apiKey',
        //   authDomain: 'your-firebase-authDomain',
        //   databaseURL: 'your-firebase-databaseURL',
        //   projectId: 'your-firebase-projectId',
        //   storageBucket: 'your-firebase-storageBucket',
        //   messagingSenderId: 'your-firebase-messagingSenderId'`;
        const moduleName = `NgxAuthFirebaseUIModule.forRoot(PUT_YOUR_FIREBASE_API_KEY_HERE)`;
        schematics_2.addModuleImportToRootModule(host, moduleName, 'ngx-auth-firebaseui', project);
        context.logger.log('info', `✅️ "${moduleName}" is imported`);
        return host;
    };
}
exports.addModuleToImports = addModuleToImports;
/** Gets the version of the specified package by looking at the package.json in the given tree. */
function getPackageVersionFromPackageJson(tree, name) {
    if (!tree.exists('package.json')) {
        return null;
    }
    // tslint:disable-next-line:no-non-null-assertion
    const packageJson = JSON.parse(tree.read('package.json').toString('utf8'));
    if (packageJson.dependencies && packageJson.dependencies[name]) {
        return packageJson.dependencies[name];
    }
    return null;
}
exports.getPackageVersionFromPackageJson = getPackageVersionFromPackageJson;
function addLibAssetsToAssets(options) {
    return (host, context) => {
        const ngxAuthFirebaseui = 'ngx-auth-firebaseui';
        const assetPath = 'node_modules/ngx-auth-firebaseui/assets/';
        try {
            const angularJsonFile = host.read('angular.json');
            if (angularJsonFile) {
                const angularJsonFileObject = JSON.parse(angularJsonFile.toString('utf-8'));
                const project = options.project ? options.project : Object.keys(angularJsonFileObject.projects)[0];
                const projectObject = angularJsonFileObject.projects[project];
                const assets = projectObject.architect.build.options.assets;
                context.logger.log('info', `"${assets}`);
                assets.push({
                    glob: '**/*',
                    input: assetPath,
                    output: './assets/'
                });
                host.overwrite('angular.json', JSON.stringify(angularJsonFileObject, null, 2));
                context.logger.log('info', `✅️ Added "${ngxAuthFirebaseui}" icons to assets`);
            }
        }
        catch (e) {
            context.logger.log('error', `🚫 Failed to add the icons "${ngxAuthFirebaseui}" to assets`);
            context.logger.log('error', e);
        }
        return host;
    };
}
function default_1(options) {
    return schematics_1.chain([
        options && options.skipPackageJson ? schematics_1.noop() : addPackageJsonDependencies(),
        options && options.skipPackageJson ? schematics_1.noop() : installPackageJsonDependencies(),
        options && options.skipModuleImport ? schematics_1.noop() : addModuleToImports(options),
        options && options.skipPolyfill ? schematics_1.noop() : addLibAssetsToAssets(options)
    ]);
}
exports.default = default_1;
