// ngx-auth-firebaseui/esm2015/lib/guards/logged-in.guard.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot } from '@angular/router';
import { map } from 'rxjs/operators';
import { NgxAuthFirebaseUIConfigToken } from '../tokens';
import { AuthProcessService } from '../services/auth-process.service';
import * as i0 from "@angular/core";
import * as i1 from "../tokens/index";
import * as i2 from "@angular/router";
import * as i3 from "../services/auth-process.service";
let LoggedInGuard = class LoggedInGuard {
    constructor(config, router, authProcess) {
        this.config = config;
        this.router = router;
        this.authProcess = authProcess;
    }
    canActivate(route, state) {
        return this.authProcess.afa.user.pipe(map(user => {
            if (user) {
                if (this.config.guardProtectedRoutesUntilEmailIsVerified && !user.emailVerified && !user.isAnonymous) {
                    if (this.config.authGuardFallbackURL) {
                        this.router.navigate([`${this.config.authGuardFallbackURL}`], { queryParams: { redirectUrl: state.url } });
                    }
                    return false;
                }
                else {
                    return true;
                }
            }
            else {
                if (this.config.authGuardFallbackURL) {
                    this.router.navigate([`/${this.config.authGuardFallbackURL}`], { queryParams: { redirectUrl: state.url } });
                }
                return false;
            }
        }));
    }
};
LoggedInGuard.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NgxAuthFirebaseUIConfigToken,] }] },
    { type: Router },
    { type: AuthProcessService }
];
LoggedInGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function LoggedInGuard_Factory() { return new LoggedInGuard(i0.ɵɵinject(i1.NgxAuthFirebaseUIConfigToken), i0.ɵɵinject(i2.Router), i0.ɵɵinject(i3.AuthProcessService)); }, token: LoggedInGuard, providedIn: "root" });
LoggedInGuard = __decorate([
    Injectable({
        providedIn: 'root'
    }),
    __param(0, Inject(NgxAuthFirebaseUIConfigToken))
], LoggedInGuard);
export { LoggedInGuard };
