// ngx-auth-firebaseui/esm5/lib/guards/logged-in.guard.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot } from '@angular/router';
import { map } from 'rxjs/operators';
import { NgxAuthFirebaseUIConfigToken } from '../tokens';
import { AuthProcessService } from '../services/auth-process.service';
import * as i0 from "@angular/core";
import * as i1 from "../tokens/index";
import * as i2 from "@angular/router";
import * as i3 from "../services/auth-process.service";
var LoggedInGuard = /** @class */ (function () {
    function LoggedInGuard(config, router, authProcess) {
        this.config = config;
        this.router = router;
        this.authProcess = authProcess;
    }
    LoggedInGuard.prototype.canActivate = function (route, state) {
        var _this = this;
        return this.authProcess.afa.user.pipe(map(function (user) {
            if (user) {
                if (_this.config.guardProtectedRoutesUntilEmailIsVerified && !user.emailVerified && !user.isAnonymous) {
                    if (_this.config.authGuardFallbackURL) {
                        _this.router.navigate(["" + _this.config.authGuardFallbackURL], { queryParams: { redirectUrl: state.url } });
                    }
                    return false;
                }
                else {
                    return true;
                }
            }
            else {
                if (_this.config.authGuardFallbackURL) {
                    _this.router.navigate(["/" + _this.config.authGuardFallbackURL], { queryParams: { redirectUrl: state.url } });
                }
                return false;
            }
        }));
    };
    LoggedInGuard.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [NgxAuthFirebaseUIConfigToken,] }] },
        { type: Router },
        { type: AuthProcessService }
    ]; };
    LoggedInGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function LoggedInGuard_Factory() { return new LoggedInGuard(i0.ɵɵinject(i1.NgxAuthFirebaseUIConfigToken), i0.ɵɵinject(i2.Router), i0.ɵɵinject(i3.AuthProcessService)); }, token: LoggedInGuard, providedIn: "root" });
    LoggedInGuard = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __param(0, Inject(NgxAuthFirebaseUIConfigToken))
    ], LoggedInGuard);
    return LoggedInGuard;
}());
export { LoggedInGuard };
