// ngx-auth-firebaseui/esm5/lib/components/ngx-auth-firebaseui-avatar/ngx-auth-firebaseui-avatar.component.js
import { __awaiter, __decorate, __generator } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { AngularFireAuth } from '@angular/fire/auth';
import { MatDialog } from '@angular/material/dialog';
import { UserComponent } from '..';
var NgxAuthFirebaseuiAvatarComponent = /** @class */ (function () {
    function NgxAuthFirebaseuiAvatarComponent(afa, dialog) {
        this.afa = afa;
        this.dialog = dialog;
        this.layout = 'default';
        this.canLogout = true;
        this.canViewAccount = true;
        this.canDeleteAccount = true;
        this.canEditAccount = true;
        this.textProfile = 'Profile';
        this.textSignOut = 'Sign Out';
        // tslint:disable-next-line:no-output-on-prefix
        this.onSignOut = new EventEmitter();
    }
    NgxAuthFirebaseuiAvatarComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.user$ = this.afa.user;
        this.user$.subscribe(function (user) {
            _this.user = user;
            _this.displayNameInitials = user ? _this.getDisplayNameInitials(user.displayName) : null;
        });
    };
    NgxAuthFirebaseuiAvatarComponent.prototype.getDisplayNameInitials = function (displayName) {
        if (!displayName) {
            return null;
        }
        var initialsRegExp = displayName.match(/\b\w/g) || [];
        var initials = ((initialsRegExp.shift() || '') + (initialsRegExp.pop() || '')).toUpperCase();
        return initials;
    };
    NgxAuthFirebaseuiAvatarComponent.prototype.openProfile = function () {
        var dialogRef = this.dialog.open(UserComponent);
        var instance = dialogRef.componentInstance;
        instance.canDeleteAccount = this.canDeleteAccount;
        instance.canEditAccount = this.canEditAccount;
    };
    NgxAuthFirebaseuiAvatarComponent.prototype.signOut = function () {
        return __awaiter(this, void 0, void 0, function () {
            var e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.afa.signOut()];
                    case 1:
                        _a.sent();
                        // Sign-out successful.
                        this.onSignOut.emit();
                        return [3 /*break*/, 3];
                    case 2:
                        e_1 = _a.sent();
                        // An error happened.
                        console.error('An error happened while signing out!', e_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    NgxAuthFirebaseuiAvatarComponent.ctorParameters = function () { return [
        { type: AngularFireAuth },
        { type: MatDialog }
    ]; };
    __decorate([
        Input()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "layout", void 0);
    __decorate([
        Input()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "canLogout", void 0);
    __decorate([
        Input()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "links", void 0);
    __decorate([
        Input()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "canViewAccount", void 0);
    __decorate([
        Input()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "canDeleteAccount", void 0);
    __decorate([
        Input()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "canEditAccount", void 0);
    __decorate([
        Input()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "textProfile", void 0);
    __decorate([
        Input()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "textSignOut", void 0);
    __decorate([
        Output()
    ], NgxAuthFirebaseuiAvatarComponent.prototype, "onSignOut", void 0);
    NgxAuthFirebaseuiAvatarComponent = __decorate([
        Component({
            selector: 'ngx-auth-firebaseui-avatar',
            template: "<button *ngIf=\"user\"\n        class=\"ngx-auth-firebaseui-avatar-button\"\n        [matMenuTriggerFor]=\"posXMenu\"\n        [matTooltip]=\"user?.displayName\"\n        [style.background-image]=\"'url(' + user?.photoURL + ')'\"\n        aria-label=\"Open x-positioned menu\"\n        mat-mini-fab\n        style=\"background-size: cover\">\n  <span *ngIf=\"!user?.photoURL\">{{displayNameInitials || ''}}</span>\n</button>\n\n<mat-menu #posXMenu=\"matMenu\" class=\"before ngx-auth-firebaseui-avatar-menu\" xPosition=\"before\" >\n  <div fxLayout=\"row\" fxLayout.xs=\"column\" style=\"padding-left: 10px; padding-right: 10px\" [ngStyle]=\"{ 'padding-top.px': layout === 'default' ? 0 : 10 }\">\n    <button [style.background-image]=\"'url(' + user?.photoURL + ')'\"\n            mat-fab\n            style=\"background-size: cover\"\n            *ngIf=\"layout === 'default'\">\n      <span *ngIf=\"!user?.photoURL\">{{displayNameInitials || ''}}</span>\n    </button>\n    <div fxLayout=\"column\" style=\"padding-left: 10px; padding-right: 10px\">\n      <strong mat-card-title>{{user?.displayName}}</strong>\n      <em mat-card-subtitle style=\"font-style: italic\">{{user?.email}}</em>\n    </div>\n  </div>\n\n  <div fxFlex=\"100\" fxLayout=\"column\" [ngStyle]=\"{ 'padding-bottom.px': layout === 'default' ? 0 : 10 } \">\n    <div *ngFor=\"let menuItem of links\" class=\"links-menu\">\n      <button (click)=\"menuItem?.callback()\" mat-menu-item>\n        <mat-icon>{{menuItem?.icon}}</mat-icon>\n        {{menuItem?.text}}</button>\n    </div>\n    <button *ngIf=\"canViewAccount\" (click)=\"openProfile()\" color=\"primary\" fxLayoutAlign=\"center\" mat-raised-button>{{ textProfile }}\n    </button>\n    <button (click)=\"signOut()\" *ngIf=\"canLogout\" color=\"warn\" fxLayoutAlign=\"center\" mat-raised-button>{{ textSignOut }}\n    </button>\n  </div>\n</mat-menu>\n",
            styles: [".mat-raised-button{margin:.2rem 1rem}.links-menu{text-align:center}"]
        })
    ], NgxAuthFirebaseuiAvatarComponent);
    return NgxAuthFirebaseuiAvatarComponent;
}());
export { NgxAuthFirebaseuiAvatarComponent };
