// ngx-auth-firebaseui/esm5/lib/services/firestore-sync.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { AngularFirestore, AngularFirestoreDocument } from '@angular/fire/firestore';
import * as i0 from "@angular/core";
import * as i1 from "@angular/fire/firestore";
export var collections = {
    users: 'users',
};
var FirestoreSyncService = /** @class */ (function () {
    function FirestoreSyncService(afs) {
        this.afs = afs;
        // this.afs.firestore.settings({timestampsInSnapshots: true});
    }
    // get timestamp() {
    //     return firebase.firestore.FieldValue.serverTimestamp();
    // }
    FirestoreSyncService.prototype.getUserDocRefByUID = function (uid) {
        return this.afs.doc(collections.users + "/" + uid);
    };
    FirestoreSyncService.prototype.deleteUserData = function (uid) {
        var userRef = this.getUserDocRefByUID(uid);
        return userRef.delete();
    };
    FirestoreSyncService.prototype.updateUserData = function (user) {
        // Sets user$ data to firestore on login
        var userRef = this.getUserDocRefByUID(user.uid);
        var data = {
            uid: user.uid,
            email: user.email,
            displayName: user.displayName,
            photoURL: user.photoURL,
            phoneNumber: user.phoneNumber,
            providerId: user.providerId
        };
        return userRef.set(data, { merge: true });
    };
    FirestoreSyncService.ctorParameters = function () { return [
        { type: AngularFirestore }
    ]; };
    FirestoreSyncService.ɵprov = i0.ɵɵdefineInjectable({ factory: function FirestoreSyncService_Factory() { return new FirestoreSyncService(i0.ɵɵinject(i1.AngularFirestore)); }, token: FirestoreSyncService, providedIn: "root" });
    FirestoreSyncService = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], FirestoreSyncService);
    return FirestoreSyncService;
}());
export { FirestoreSyncService };
