// ngx-auth-firebaseui/esm5/lib/components/legality-dialog/legality-dialog.component.js
import { __decorate, __param } from "tslib";
import { Component, Inject } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
var LegalityDialogComponent = /** @class */ (function () {
    function LegalityDialogComponent(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.data = data;
        // tslint:disable-next-line:variable-name
        this._disableConfirmActionButton = false;
    }
    Object.defineProperty(LegalityDialogComponent.prototype, "disableConfirmActionButton", {
        get: function () {
            if (this.data.tosUrl && this.data.privacyPolicyUrl) {
                this._disableConfirmActionButton = !(this.checkTOS && this.checkPrivacyPolicy);
            }
            else if (this.data.tosUrl && !this.data.privacyPolicyUrl) {
                this._disableConfirmActionButton = !this.checkTOS;
            }
            else if (!this.data.tosUrl && this.data.privacyPolicyUrl) {
                this._disableConfirmActionButton = !this.checkPrivacyPolicy;
            }
            return this._disableConfirmActionButton;
        },
        enumerable: true,
        configurable: true
    });
    LegalityDialogComponent.prototype.closeDialog = function () {
        var result = {
            checked: !this.disableConfirmActionButton,
            authProvider: this.data.authProvider
        };
        this.dialogRef.close(result);
    };
    LegalityDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    LegalityDialogComponent = __decorate([
        Component({
            selector: 'ngx-auth-firebaseui-legality-dialog',
            template: "<h1 matDialogTitle>Legal requirements</h1>\n\n<mat-dialog-content>\n  <div fxLayout=\"column\" fxLayoutAlign=\"start\">\n    <mat-checkbox *ngIf=\"this.data.tosUrl\" [(ngModel)]=\"checkTOS\">\n      I agree to the\n      <span>&nbsp;</span>\n      <a [href]=\"this.data.tosUrl\"\n         target=\"_blank\">\n        Terms of Service and Conditions\n      </a>\n    </mat-checkbox>\n\n    <mat-checkbox *ngIf=\"this.data.privacyPolicyUrl\"\n                  [(ngModel)]=\"checkPrivacyPolicy\">\n      I have read and agree to the\n      <span>&nbsp;</span>\n      <a [href]=\"this.data.privacyPolicyUrl\"\n         target=\"_blank\">\n        Privacy\n      </a>\n    </mat-checkbox>\n  </div>\n</mat-dialog-content>\n\n<mat-dialog-actions>\n  <button color=\"warn\"\n          id=\"decline-action\"\n          mat-raised-button\n          matDialogClose>Decline\n  </button>\n  <button (click)=\"closeDialog()\"\n          [disabled]=\"disableConfirmActionButton\"\n          color=\"primary\"\n          id=\"confirm-action\"\n          mat-raised-button>Confirm\n  </button>\n</mat-dialog-actions>\n\n",
            styles: ["::ng-deep .mat-checkbox-label{display:flex;flex-wrap:wrap}mat-dialog-content div{margin-top:1.5rem}mat-dialog-actions{margin-top:1rem}"]
        }),
        __param(1, Inject(MAT_DIALOG_DATA))
    ], LegalityDialogComponent);
    return LegalityDialogComponent;
}());
export { LegalityDialogComponent };
