// ngx-auth-firebaseui/bundles/ngx-auth-firebaseui.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/router'), require('@angular/fire/auth'), require('@firebase/app'), require('@firebase/auth'), require('rxjs'), require('rxjs/operators'), require('@angular/fire/firestore'), require('@angular/material/snack-bar'), require('@angular/material/dialog'), require('@angular/common'), require('@angular/forms'), require('@angular/material/tabs'), require('@angular-material-extensions/password-strength'), require('@angular/animations'), require('@angular/common/http'), require('@angular/flex-layout'), require('@angular/fire'), require('@angular/material/button'), require('@angular/material/card'), require('@angular/material/checkbox'), require('@angular/material/chips'), require('@angular/material/divider'), require('@angular/material/icon'), require('@angular/material/input'), require('@angular/material/menu'), require('@angular/material/progress-bar'), require('@angular/material/progress-spinner'), require('@angular/material/tooltip'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define('ngx-auth-firebaseui', ['exports', '@angular/core', '@angular/router', '@angular/fire/auth', '@firebase/app', '@firebase/auth', 'rxjs', 'rxjs/operators', '@angular/fire/firestore', '@angular/material/snack-bar', '@angular/material/dialog', '@angular/common', '@angular/forms', '@angular/material/tabs', '@angular-material-extensions/password-strength', '@angular/animations', '@angular/common/http', '@angular/flex-layout', '@angular/fire', '@angular/material/button', '@angular/material/card', '@angular/material/checkbox', '@angular/material/chips', '@angular/material/divider', '@angular/material/icon', '@angular/material/input', '@angular/material/menu', '@angular/material/progress-bar', '@angular/material/progress-spinner', '@angular/material/tooltip', '@angular/platform-browser'], factory) :
    (global = global || self, factory(global['ngx-auth-firebaseui'] = {}, global.ng.core, global.ng.router, global.ng.fire.auth, global.app, null, global.rxjs, global.rxjs.operators, global.ng.fire.firestore, global.ng.material.snackBar, global.ng.material.dialog, global.ng.common, global.ng.forms, global.ng.material.tabs, global.passwordStrength, global.ng.animations, global.ng.common.http, global.ng.flexLayout, global.ng.fire, global.ng.material.button, global.ng.material.card, global.ng.material.checkbox, global.ng.material.chips, global.ng.material.divider, global.ng.material.icon, global.ng.material.input, global.ng.material.menu, global.ng.material.progressBar, global.ng.material.progressSpinner, global.ng.material.tooltip, global.ng.platformBrowser));
}(this, (function (exports, core, router, auth, app, auth$1, rxjs, operators, firestore, snackBar, dialog, common, forms, tabs, passwordStrength, animations, http, flexLayout, fire, button, card, checkbox, chips, divider, icon, input, menu, progressBar, progressSpinner, tooltip, platformBrowser) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    (function (Accounts) {
        Accounts["NONE"] = "account";
        Accounts["CHECK"] = "account-check";
        Accounts["EDIT"] = "account-edit";
        Accounts["OFF"] = "account-off";
        Accounts["REMOVE"] = "account-remove";
    })(exports.Accounts || (exports.Accounts = {}));

    var collections = {
        users: 'users',
    };
    var FirestoreSyncService = /** @class */ (function () {
        function FirestoreSyncService(afs) {
            this.afs = afs;
            // this.afs.firestore.settings({timestampsInSnapshots: true});
        }
        // get timestamp() {
        //     return firebase.firestore.FieldValue.serverTimestamp();
        // }
        FirestoreSyncService.prototype.getUserDocRefByUID = function (uid) {
            return this.afs.doc(collections.users + "/" + uid);
        };
        FirestoreSyncService.prototype.deleteUserData = function (uid) {
            var userRef = this.getUserDocRefByUID(uid);
            return userRef.delete();
        };
        FirestoreSyncService.prototype.updateUserData = function (user) {
            // Sets user$ data to firestore on login
            var userRef = this.getUserDocRefByUID(user.uid);
            var data = {
                uid: user.uid,
                email: user.email,
                displayName: user.displayName,
                photoURL: user.photoURL,
                phoneNumber: user.phoneNumber,
                providerId: user.providerId
            };
            return userRef.set(data, { merge: true });
        };
        FirestoreSyncService.ctorParameters = function () { return [
            { type: firestore.AngularFirestore }
        ]; };
        FirestoreSyncService.ɵprov = core.ɵɵdefineInjectable({ factory: function FirestoreSyncService_Factory() { return new FirestoreSyncService(core.ɵɵinject(firestore.AngularFirestore)); }, token: FirestoreSyncService, providedIn: "root" });
        FirestoreSyncService = __decorate([
            core.Injectable({
                providedIn: 'root'
            })
        ], FirestoreSyncService);
        return FirestoreSyncService;
    }());

    // This token is the official token containing the final configuration; ie. the merge between default and user provided configurations
    var NgxAuthFirebaseUIConfigToken = new core.InjectionToken('NgxAuthFirebaseUIConfigToken');
    // This is an intermediate token containing only user-provided configuration
    var UserProvidedConfigToken = new core.InjectionToken('UserProvidedConfigToken');

    var facebookAuthProvider = new app.firebase.auth.FacebookAuthProvider();
    var googleAuthProvider = new app.firebase.auth.GoogleAuthProvider();
    var appleAuthProvider = new app.firebase.auth.OAuthProvider('apple.com');
    var twitterAuthProvider = new app.firebase.auth.TwitterAuthProvider();
    var githubAuthProvider = new app.firebase.auth.GithubAuthProvider();
    var microsoftAuthProvider = new app.firebase.auth.OAuthProvider('microsoft.com');
    var yahooAuthProvider = new app.firebase.auth.OAuthProvider('yahoo.com');

    (function (AuthProvider) {
        AuthProvider["ALL"] = "all";
        AuthProvider["ANONYMOUS"] = "anonymous";
        AuthProvider["EmailAndPassword"] = "firebase";
        AuthProvider["Google"] = "google";
        AuthProvider["Apple"] = "apple";
        AuthProvider["Facebook"] = "facebook";
        AuthProvider["Twitter"] = "twitter";
        AuthProvider["Github"] = "github";
        AuthProvider["Microsoft"] = "microsoft";
        AuthProvider["Yahoo"] = "yahoo";
        AuthProvider["PhoneNumber"] = "phoneNumber";
    })(exports.AuthProvider || (exports.AuthProvider = {}));
    var AuthProcessService = /** @class */ (function () {
        function AuthProcessService(afa, config, snackBar, fireStoreService, matSnackBarConfig) {
            this.afa = afa;
            this.config = config;
            this.snackBar = snackBar;
            this.fireStoreService = fireStoreService;
            this.matSnackBarConfig = matSnackBarConfig;
            this.onSuccessEmitter = new core.EventEmitter();
            this.onErrorEmitter = new core.EventEmitter();
            // Useful to know about auth state even between reloads.
            // Replace emailConfirmationSent and emailToConfirm.
            this._user$ = new rxjs.BehaviorSubject(null);
        }
        Object.defineProperty(AuthProcessService.prototype, "user$", {
            get: function () {
                return this._user$.asObservable();
            },
            enumerable: true,
            configurable: true
        });
        AuthProcessService.prototype.listenToUserEvents = function () {
            var _this = this;
            this.afa.user.subscribe(function (user) {
                _this._user$.next(user);
                _this.user = user;
            });
        };
        /**
         * Reset the password of the ngx-auth-firebaseui-user via email
         *
         * @param email - the email to reset
         */
        AuthProcessService.prototype.resetPassword = function (email) {
            return __awaiter(this, void 0, void 0, function () {
                var error_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 2, , 3]);
                            console.log('Password reset email sent');
                            return [4 /*yield*/, this.afa.sendPasswordResetEmail(email)];
                        case 1: return [2 /*return*/, _a.sent()];
                        case 2:
                            error_1 = _a.sent();
                            return [2 /*return*/, this.notifyError(error_1)];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * General sign in mechanism to authenticate the users with a firebase project
         * using a traditional way, via username and password or by using an authentication provider
         * like google, facebook, twitter and github
         *
         * @param provider - the provider to authenticate with (google, facebook, twitter, github)
         * @param credentials optional email and password
         */
        AuthProcessService.prototype.signInWith = function (provider, credentials) {
            return __awaiter(this, void 0, void 0, function () {
                var signInResult, _a, err_1;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            _b.trys.push([0, 23, , 24]);
                            signInResult = void 0;
                            _a = provider;
                            switch (_a) {
                                case exports.AuthProvider.ANONYMOUS: return [3 /*break*/, 1];
                                case exports.AuthProvider.EmailAndPassword: return [3 /*break*/, 3];
                                case exports.AuthProvider.Google: return [3 /*break*/, 5];
                                case exports.AuthProvider.Apple: return [3 /*break*/, 7];
                                case exports.AuthProvider.Facebook: return [3 /*break*/, 9];
                                case exports.AuthProvider.Twitter: return [3 /*break*/, 11];
                                case exports.AuthProvider.Github: return [3 /*break*/, 13];
                                case exports.AuthProvider.Microsoft: return [3 /*break*/, 15];
                                case exports.AuthProvider.Yahoo: return [3 /*break*/, 17];
                                case exports.AuthProvider.PhoneNumber: return [3 /*break*/, 19];
                            }
                            return [3 /*break*/, 20];
                        case 1: return [4 /*yield*/, this.afa.signInAnonymously()];
                        case 2:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 3: return [4 /*yield*/, this.afa.signInWithEmailAndPassword(credentials.email, credentials.password)];
                        case 4:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 5: return [4 /*yield*/, this.afa.signInWithPopup(googleAuthProvider)];
                        case 6:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 7: return [4 /*yield*/, this.afa.signInWithPopup(appleAuthProvider)];
                        case 8:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 9: return [4 /*yield*/, this.afa.signInWithPopup(facebookAuthProvider)];
                        case 10:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 11: return [4 /*yield*/, this.afa.signInWithPopup(twitterAuthProvider)];
                        case 12:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 13: return [4 /*yield*/, this.afa.signInWithPopup(githubAuthProvider)];
                        case 14:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 15: return [4 /*yield*/, this.afa.signInWithPopup(microsoftAuthProvider)];
                        case 16:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 17: return [4 /*yield*/, this.afa.signInWithPopup(yahooAuthProvider)];
                        case 18:
                            signInResult = (_b.sent());
                            return [3 /*break*/, 21];
                        case 19: 
                        // coming soon - see feature/sms branch
                        return [3 /*break*/, 21];
                        case 20: throw new Error(exports.AuthProvider[provider] + " is not available as auth provider");
                        case 21: return [4 /*yield*/, this.handleSuccess(signInResult)];
                        case 22:
                            _b.sent();
                            return [3 /*break*/, 24];
                        case 23:
                            err_1 = _b.sent();
                            this.handleError(err_1);
                            return [3 /*break*/, 24];
                        case 24: return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * Sign up new users via email and password.
         * After that the ngx-auth-firebaseui-user should verify and confirm an email sent via the firebase
         *
         * @param displayName - the displayName if the new ngx-auth-firebaseui-user
         * @param credentials email and password
         * @returns -
         */
        AuthProcessService.prototype.signUp = function (displayName, credentials) {
            return __awaiter(this, void 0, void 0, function () {
                var userCredential, user, err_2;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 8, , 9]);
                            return [4 /*yield*/, this.afa.createUserWithEmailAndPassword(credentials.email, credentials.password)];
                        case 1:
                            userCredential = _a.sent();
                            user = userCredential.user;
                            return [4 /*yield*/, this.updateProfile(displayName, user.photoURL)];
                        case 2:
                            _a.sent();
                            if (!this.config.enableFirestoreSync) return [3 /*break*/, 4];
                            return [4 /*yield*/, this.fireStoreService
                                    .getUserDocRefByUID(user.uid)
                                    .set({
                                    uid: user.uid,
                                    displayName: displayName,
                                    email: user.email,
                                    photoURL: user.photoURL
                                })];
                        case 3:
                            _a.sent();
                            _a.label = 4;
                        case 4:
                            if (!this.config.enableEmailVerification) return [3 /*break*/, 6];
                            return [4 /*yield*/, user.sendEmailVerification()];
                        case 5:
                            _a.sent();
                            _a.label = 6;
                        case 6:
                            // Legacy fields
                            this.emailConfirmationSent = true;
                            this.emailToConfirm = credentials.email;
                            return [4 /*yield*/, this.handleSuccess(userCredential)];
                        case 7:
                            _a.sent();
                            return [3 /*break*/, 9];
                        case 8:
                            err_2 = _a.sent();
                            this.handleError(err_2);
                            return [3 /*break*/, 9];
                        case 9: return [2 /*return*/];
                    }
                });
            });
        };
        AuthProcessService.prototype.sendNewVerificationEmail = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    if (!this.user) {
                        return [2 /*return*/, Promise.reject(new Error('No signed in user'))];
                    }
                    return [2 /*return*/, this.user.sendEmailVerification()];
                });
            });
        };
        AuthProcessService.prototype.signOut = function () {
            return __awaiter(this, void 0, void 0, function () {
                var error_2;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 2, , 3]);
                            return [4 /*yield*/, this.afa.signOut()];
                        case 1:
                            _a.sent();
                            return [3 /*break*/, 3];
                        case 2:
                            error_2 = _a.sent();
                            this.notifyError(error_2);
                            return [3 /*break*/, 3];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * Update the profile (name + photo url) of the authenticated ngx-auth-firebaseui-user in the
         * firebase authentication feature (not in firestore)
         *
         * @param name - the new name of the authenticated ngx-auth-firebaseui-user
         * @param photoURL - the new photo url of the authenticated ngx-auth-firebaseui-user
         * @returns -
         */
        AuthProcessService.prototype.updateProfile = function (name, photoURL) {
            return this.afa.currentUser.then(function (user) {
                if (!photoURL) {
                    return user.updateProfile({ displayName: name });
                }
                else {
                    return user.updateProfile({ displayName: name, photoURL: photoURL });
                }
            });
        };
        AuthProcessService.prototype.parseUserInfo = function (user) {
            return {
                uid: user.uid,
                displayName: user.displayName,
                email: user.email,
                phoneNumber: user.phoneNumber,
                photoURL: user.photoURL,
                providerId: user.providerData.length > 0 ? user.providerData[0].providerId : null
            };
        };
        AuthProcessService.prototype.getUserPhotoUrl = function () {
            var _this = this;
            return this._user$.pipe(operators.map(function (user) {
                if (!user) {
                    return null;
                }
                else if (user.photoURL) {
                    return user.photoURL;
                }
                else if (user.emailVerified) {
                    return _this.getPhotoPath(exports.Accounts.CHECK);
                }
                else if (user.isAnonymous) {
                    return _this.getPhotoPath(exports.Accounts.OFF);
                }
                else {
                    return _this.getPhotoPath(exports.Accounts.NONE);
                }
            }));
        };
        AuthProcessService.prototype.getPhotoPath = function (image) {
            return "assets/user/" + image + ".svg";
        };
        AuthProcessService.prototype.signInWithPhoneNumber = function () {
            // todo: 3.1.18
        };
        AuthProcessService.prototype.handleSuccess = function (userCredential) {
            return __awaiter(this, void 0, void 0, function () {
                var e_1, fallbackMessage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            this.onSuccessEmitter.next(userCredential.user);
                            if (!this.config.enableFirestoreSync) return [3 /*break*/, 4];
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, this.fireStoreService.updateUserData(this.parseUserInfo(userCredential.user))];
                        case 2:
                            _a.sent();
                            return [3 /*break*/, 4];
                        case 3:
                            e_1 = _a.sent();
                            console.error("Error occurred while updating user data with firestore: " + e_1);
                            return [3 /*break*/, 4];
                        case 4:
                            if (this.config.toastMessageOnAuthSuccess) {
                                fallbackMessage = "Hello " + (userCredential.user.displayName ? userCredential.user.displayName : '') + "!";
                                this.showToast(this.messageOnAuthSuccess || fallbackMessage);
                            }
                            return [2 /*return*/];
                    }
                });
            });
        };
        AuthProcessService.prototype.handleError = function (error) {
            this.notifyError(error);
            console.error(error);
        };
        // Refresh user info. Can be useful for instance to get latest status regarding email verification.
        AuthProcessService.prototype.reloadUserInfo = function () {
            return this._user$.pipe(operators.take(1)).subscribe(function (user) { return user && user.reload(); });
        };
        // Search for an error message.
        // Consumers of this library are given the possibility to provide a
        // function in case they want to instrument message based on error properties.
        AuthProcessService.prototype.getMessageOnAuthError = function (error) {
            // tslint:disable-next-line:no-bitwise
            return error.toString() || 'Sorry, something went wrong. Please retry later.';
        };
        // Show a toast using current snackbar config. If message is empty, no toast is displayed allowing to opt-out when needed.
        // Default MatSnackBarConfig has no duration, meaning it stays visible forever.
        // If that's the case, an action button is added to allow the end-user to dismiss the toast.
        AuthProcessService.prototype.showToast = function (message) {
            if (message) {
                this.snackBar.open(message, this.matSnackBarConfig.duration ? null : 'OK');
            }
        };
        AuthProcessService.prototype.showErrorToast = function (error) {
            if (this.config.toastMessageOnAuthError) {
                this.showToast(this.getMessageOnAuthError(error));
            }
        };
        AuthProcessService.prototype.notifyError = function (error) {
            this.onErrorEmitter.emit(error);
            this.showErrorToast(error);
        };
        AuthProcessService.ctorParameters = function () { return [
            { type: auth.AngularFireAuth },
            { type: undefined, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return NgxAuthFirebaseUIConfigToken; }),] }] },
            { type: snackBar.MatSnackBar },
            { type: FirestoreSyncService },
            { type: snackBar.MatSnackBarConfig, decorators: [{ type: core.Inject, args: [snackBar.MAT_SNACK_BAR_DEFAULT_OPTIONS,] }] }
        ]; };
        AuthProcessService.ɵprov = core.ɵɵdefineInjectable({ factory: function AuthProcessService_Factory() { return new AuthProcessService(core.ɵɵinject(auth.AngularFireAuth), core.ɵɵinject(NgxAuthFirebaseUIConfigToken), core.ɵɵinject(snackBar.MatSnackBar), core.ɵɵinject(FirestoreSyncService), core.ɵɵinject(snackBar.MAT_SNACK_BAR_DEFAULT_OPTIONS)); }, token: AuthProcessService, providedIn: "root" });
        AuthProcessService = __decorate([
            core.Injectable({
                providedIn: 'root'
            }),
            __param(1, core.Inject(core.forwardRef(function () { return NgxAuthFirebaseUIConfigToken; }))),
            __param(4, core.Inject(snackBar.MAT_SNACK_BAR_DEFAULT_OPTIONS))
        ], AuthProcessService);
        return AuthProcessService;
    }());

    var defaultTranslations = {
        verifyEmailTitleText: 'Confirm your e-mail address!',
        verifyEmailConfirmationText: 'A confirmation e-mail has been sent.' +
            ' Check your inbox and click on the link "Confirm my e-mail" to confirm your e-mail address.',
        verifyEmailGoBackText: 'Go back',
        sendNewVerificationEmailText: 'Send new confirmation e-mail',
        signOutText: 'Sign out',
        messageOnEmailConfirmationSuccess: 'A new confirmation e-mail has been sent. Please check your inbox.',
    };
    var EmailConfirmationComponent = /** @class */ (function () {
        function EmailConfirmationComponent(authProcess, router, changeDetectorRef) {
            this.authProcess = authProcess;
            this.router = router;
            this.changeDetectorRef = changeDetectorRef;
            this.signOut = new core.EventEmitter();
        }
        EmailConfirmationComponent.prototype.ngOnChanges = function (changes) {
            if (changes.verifyEmailTemplate && changes.verifyEmailTemplate.currentValue == null) {
                this.verifyEmailTemplate = this.defaultTemplate;
                console.log('ngOnChanges - defaultTemplate:', this.verifyEmailTemplate);
            }
            this.verifyEmailContext = this.createTemplateContext();
        };
        EmailConfirmationComponent.prototype.ngOnInit = function () {
            if (!this.verifyEmailTemplate) {
                console.log('ngOnInit - defaultTemplate');
                this.verifyEmailTemplate = this.defaultTemplate;
            }
            this.verifyEmailContext = this.createTemplateContext();
            console.log('verifyEmailTemplate:', this.verifyEmailTemplate);
            console.log('verifyEmailContext:', this.verifyEmailContext);
        };
        EmailConfirmationComponent.prototype.continue = function () {
            return __awaiter(this, void 0, void 0, function () {
                var error_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 3, , 4]);
                            return [4 /*yield*/, this.authProcess.reloadUserInfo()];
                        case 1:
                            _a.sent();
                            return [4 /*yield*/, this.router.navigate([this.goBackURL])];
                        case 2:
                            _a.sent();
                            return [3 /*break*/, 4];
                        case 3:
                            error_1 = _a.sent();
                            this.authProcess.notifyError(error_1);
                            return [3 /*break*/, 4];
                        case 4: return [2 /*return*/];
                    }
                });
            });
        };
        EmailConfirmationComponent.prototype.sendNewVerificationEmail = function () {
            return __awaiter(this, void 0, void 0, function () {
                var error_2;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 2, 3, 4]);
                            this.isLoading = true;
                            this.changeDetectorRef.markForCheck();
                            return [4 /*yield*/, this.authProcess.sendNewVerificationEmail()];
                        case 1:
                            _a.sent();
                            this.authProcess.showToast(this.verifyEmailContext.messageOnEmailConfirmationSuccess);
                            return [3 /*break*/, 4];
                        case 2:
                            error_2 = _a.sent();
                            this.authProcess.notifyError(error_2);
                            return [3 /*break*/, 4];
                        case 3:
                            this.isLoading = false;
                            this.changeDetectorRef.markForCheck();
                            return [7 /*endfinally*/];
                        case 4: return [2 /*return*/];
                    }
                });
            });
        };
        EmailConfirmationComponent.prototype.createTemplateContext = function () {
            return {
                email: this.email,
                goBackURL: this.goBackURL,
                verifyEmailTitleText: this.verifyEmailTitleText || defaultTranslations.verifyEmailTitleText,
                verifyEmailConfirmationText: this.verifyEmailConfirmationText || defaultTranslations.verifyEmailConfirmationText,
                verifyEmailGoBackText: this.verifyEmailGoBackText || defaultTranslations.verifyEmailGoBackText,
                sendNewVerificationEmailText: this.sendNewVerificationEmailText || defaultTranslations.sendNewVerificationEmailText,
                signOutText: this.signOutText || defaultTranslations.signOutText,
                messageOnEmailConfirmationSuccess: this.messageOnEmailConfirmationSuccess || defaultTranslations.messageOnEmailConfirmationSuccess
            };
        };
        EmailConfirmationComponent.ctorParameters = function () { return [
            { type: AuthProcessService },
            { type: router.Router },
            { type: core.ChangeDetectorRef }
        ]; };
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "email", void 0);
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "goBackURL", void 0);
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "verifyEmailTitleText", void 0);
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "verifyEmailConfirmationText", void 0);
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "verifyEmailGoBackText", void 0);
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "sendNewVerificationEmailText", void 0);
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "signOutText", void 0);
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "messageOnEmailConfirmationSuccess", void 0);
        __decorate([
            core.Input()
        ], EmailConfirmationComponent.prototype, "template", void 0);
        __decorate([
            core.Output()
        ], EmailConfirmationComponent.prototype, "signOut", void 0);
        __decorate([
            core.ViewChild('defaultVerifyEmail', { static: true })
        ], EmailConfirmationComponent.prototype, "defaultTemplate", void 0);
        EmailConfirmationComponent = __decorate([
            core.Component({
                selector: 'ngx-auth-firebaseui-email-confirmation',
                template: "<ng-container *ngTemplateOutlet=\"verifyEmailTemplate; context: verifyEmailContext\"></ng-container>\n<ng-template #defaultVerifyEmail let-email=\"email\" let-goBackURL=\"goBackURL\"\n             let-sendNewVerificationEmailText=\"sendNewVerificationEmailText\"\n             let-signOutText=\"signOutText\"\n             let-verifyEmailConfirmationText=\"verifyEmailConfirmationText\" let-verifyEmailGoBackText=\"verifyEmailGoBackText\"\n             let-verifyEmailTitleText=\"verifyEmailTitleText\">\n  <mat-card class=\"verify-email\">\n    <mat-card-content fxLayout=\"column\" fxLayoutAlign=\"center center\">\n      <mat-icon>email</mat-icon>\n      <p class=\"title\" fxLayout=\"column\" fxLayoutAlign=\"center center\">\n        <span class=\"mat-subheading-2\">{{ verifyEmailTitleText }}</span>\n        <span class=\"mat-body-2\">{{ email }}</span>\n      </p>\n      <p class=\"subtitle\">{{ verifyEmailConfirmationText }}</p>\n      <mat-progress-bar *ngIf=\"isLoading\" mode=\"indeterminate\"></mat-progress-bar>\n    </mat-card-content>\n    <mat-card-actions fxLayout=\"column\" fxLayoutAlign=\"center center\">\n      <button (click)=\"continue()\" *ngIf=\"goBackURL\" class=\"go-back-button action-button\" mat-stroked-button>\n        {{ verifyEmailGoBackText }}\n      </button>\n      <button (click)=\"sendNewVerificationEmail()\" class=\"send-new-mail-button action-button\"\n              mat-stroked-button>{{ sendNewVerificationEmailText }}</button>\n      <button (click)=\"this.signOut.emit()\" class=\"sign-out-button action-button\" color=\"warn\"\n              mat-stroked-button>{{ signOutText }}</button>\n    </mat-card-actions>\n  </mat-card>\n</ng-template>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [".material-icons{font-size:4rem}.verify-email{width:360px}.verify-email .mat-icon{height:4rem;width:4rem;color:#444}.verify-email .title{margin-top:16px}.verify-email .title .mat-subheading-2{margin-bottom:0}.verify-email .subtitle{margin:16px auto;text-align:justify}.verify-email p{display:block;-webkit-margin-before:1em;-webkit-margin-after:1em;-webkit-margin-start:0;-webkit-margin-end:0}.verify-email mat-card-actions{text-align:center;margin-top:1rem}.verify-email mat-card-actions .action-button{width:100%}.verify-email mat-card-actions .action-button+.action-button{margin-top:1rem}"]
            })
        ], EmailConfirmationComponent);
        return EmailConfirmationComponent;
    }());

    var LegalityDialogComponent = /** @class */ (function () {
        function LegalityDialogComponent(dialogRef, data) {
            this.dialogRef = dialogRef;
            this.data = data;
            // tslint:disable-next-line:variable-name
            this._disableConfirmActionButton = false;
        }
        Object.defineProperty(LegalityDialogComponent.prototype, "disableConfirmActionButton", {
            get: function () {
                if (this.data.tosUrl && this.data.privacyPolicyUrl) {
                    this._disableConfirmActionButton = !(this.checkTOS && this.checkPrivacyPolicy);
                }
                else if (this.data.tosUrl && !this.data.privacyPolicyUrl) {
                    this._disableConfirmActionButton = !this.checkTOS;
                }
                else if (!this.data.tosUrl && this.data.privacyPolicyUrl) {
                    this._disableConfirmActionButton = !this.checkPrivacyPolicy;
                }
                return this._disableConfirmActionButton;
            },
            enumerable: true,
            configurable: true
        });
        LegalityDialogComponent.prototype.closeDialog = function () {
            var result = {
                checked: !this.disableConfirmActionButton,
                authProvider: this.data.authProvider
            };
            this.dialogRef.close(result);
        };
        LegalityDialogComponent.ctorParameters = function () { return [
            { type: dialog.MatDialogRef },
            { type: undefined, decorators: [{ type: core.Inject, args: [dialog.MAT_DIALOG_DATA,] }] }
        ]; };
        LegalityDialogComponent = __decorate([
            core.Component({
                selector: 'ngx-auth-firebaseui-legality-dialog',
                template: "<h1 matDialogTitle>Legal requirements</h1>\n\n<mat-dialog-content>\n  <div fxLayout=\"column\" fxLayoutAlign=\"start\">\n    <mat-checkbox *ngIf=\"this.data.tosUrl\" [(ngModel)]=\"checkTOS\">\n      I agree to the\n      <span>&nbsp;</span>\n      <a [href]=\"this.data.tosUrl\"\n         target=\"_blank\">\n        Terms of Service and Conditions\n      </a>\n    </mat-checkbox>\n\n    <mat-checkbox *ngIf=\"this.data.privacyPolicyUrl\"\n                  [(ngModel)]=\"checkPrivacyPolicy\">\n      I have read and agree to the\n      <span>&nbsp;</span>\n      <a [href]=\"this.data.privacyPolicyUrl\"\n         target=\"_blank\">\n        Privacy\n      </a>\n    </mat-checkbox>\n  </div>\n</mat-dialog-content>\n\n<mat-dialog-actions>\n  <button color=\"warn\"\n          id=\"decline-action\"\n          mat-raised-button\n          matDialogClose>Decline\n  </button>\n  <button (click)=\"closeDialog()\"\n          [disabled]=\"disableConfirmActionButton\"\n          color=\"primary\"\n          id=\"confirm-action\"\n          mat-raised-button>Confirm\n  </button>\n</mat-dialog-actions>\n\n",
                styles: ["::ng-deep .mat-checkbox-label{display:flex;flex-wrap:wrap}mat-dialog-content div{margin-top:1.5rem}mat-dialog-actions{margin-top:1rem}"]
            }),
            __param(1, core.Inject(dialog.MAT_DIALOG_DATA))
        ], LegalityDialogComponent);
        return LegalityDialogComponent;
    }());

    var customAnimation = animations.animation([
        animations.style({
            opacity: '{{opacity}}',
            transform: 'scale({{scale}}) translate3d({{x}}, {{y}}, {{z}})'
        }),
        animations.animate('{{duration}} {{delay}} cubic-bezier(0.0, 0.0, 0.2, 1)', animations.style('*'))
    ], {
        params: {
            duration: '200ms',
            delay: '0ms',
            opacity: '0',
            scale: '1',
            x: '0',
            y: '0',
            z: '0'
        }
    });
    var NgxAuthFirebaseuiAnimations = [
        animations.trigger('animate', [animations.transition('void => *', [animations.useAnimation(customAnimation)])]),
        animations.trigger('animateStagger', [
            animations.state('50', animations.style('*')),
            animations.state('100', animations.style('*')),
            animations.state('200', animations.style('*')),
            animations.transition('void => 50', animations.query('@*', [animations.stagger('50ms', [animations.animateChild()])], { optional: true })),
            animations.transition('void => 100', animations.query('@*', [animations.stagger('100ms', [animations.animateChild()])], { optional: true })),
            animations.transition('void => 200', animations.query('@*', [animations.stagger('200ms', [animations.animateChild()])], { optional: true }))
        ]),
    ];

    var EMAIL_REGEX = new RegExp(['^(([^<>()[\\]\\\.,;:\\s@\"]+(\\.[^<>()\\[\\]\\\.,;:\\s@\"]+)*)',
        '|(".+"))@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.',
        '[0-9]{1,3}\])|(([a-zA-Z\\-0-9]+\\.)+',
        '[a-zA-Z]{2,}))$'].join(''));
    // tslint:disable-next-line:max-line-length
    var PHONE_NUMBER_REGEX = new RegExp(['^[+]{0,1}[(]{0,1}[0-9]{1,4}[)]{0,1}[-\\s\\.]{0,1}[(]{0,1}[0-9]{1,4}[)]{0,1}[-\\s\\./0-9]{4,12}$'].join(''));
    var AuthComponent = /** @class */ (function () {
        function AuthComponent(
        // tslint:disable-next-line:ban-types
        platformId, config, auth, authProcess, dialog, activatedRoute, changeDetectorRef) {
            this.platformId = platformId;
            this.config = config;
            this.auth = auth;
            this.authProcess = authProcess;
            this.dialog = dialog;
            this.activatedRoute = activatedRoute;
            this.changeDetectorRef = changeDetectorRef;
            //  google, facebook, twitter, github as array or all as one single string
            this.providers = exports.AuthProvider.ALL;
            this.registrationEnabled = true;
            this.resetPasswordEnabled = true;
            this.guestEnabled = true;
            this.selectedTabChange = new core.EventEmitter();
            // Password strength api
            this.enableLengthRule = true;
            this.enableLowerCaseLetterRule = true;
            this.enableUpperCaseLetterRule = true;
            this.enableDigitRule = true;
            this.enableSpecialCharRule = true;
            // tslint:disable-next-line:no-output-on-prefix
            this.onStrengthChanged = new core.EventEmitter();
            this.signOutText = 'Sign out';
            // Customize the text
            // Reset Password Tab
            this.resetPasswordTabText = 'Reset e-mail address to password';
            this.resetPasswordInputText = 'Reset e-mail address to password';
            this.resetPasswordErrorRequiredText = 'E-mail is required to reset the password!';
            this.resetPasswordErrorPatternText = 'Please enter a valid e-mail address';
            this.resetPasswordActionButtonText = 'Reset';
            this.resetPasswordInstructionsText = 'Reset requested. Check your e-mail instructions.';
            // SignIn Tab
            this.signInTabText = 'Sign in';
            this.signInCardTitleText = 'Signing in';
            this.loginButtonText = 'Log In';
            this.forgotPasswordButtonText = 'Forgot Password ?';
            // Common
            this.nameText = 'Name';
            this.nameErrorRequiredText = 'Name is required';
            this.nameErrorMinLengthText = 'The name is too short!';
            this.nameErrorMaxLengthText = 'The name is too long!';
            this.emailText = 'E-mail';
            this.emailErrorRequiredText = 'E-mail is required';
            this.emailErrorPatternText = 'Please enter a valid e-mail address';
            this.passwordText = 'Password';
            this.passwordErrorRequiredText = 'Password is required';
            this.passwordErrorMinLengthText = 'The password is too short!';
            this.passwordErrorMaxLengthText = 'The password is too long!';
            // Register Tab
            this.registerTabText = 'Register';
            this.registerCardTitleText = 'Registration';
            this.registerButtonText = 'Register';
            this.guestButtonText = 'continue as guest';
            // email confirmation component
            this.emailConfirmationTitle = 'Confirm your e-mail address!';
            // tslint:disable-next-line:max-line-length
            this.emailConfirmationText = "A confirmation e-mail has been sent to you. Check your inbox and click on the link \"Confirm my e-mail\" to confirm your e-mail address.";
            this.authProvider = exports.AuthProvider;
            this.authenticationError = false;
            this.passReset = false;
            this.authProviders = exports.AuthProvider;
            this.onSuccess = authProcess.onSuccessEmitter;
            this.onError = authProcess.onErrorEmitter;
        }
        Object.defineProperty(AuthComponent.prototype, "color", {
            get: function () {
                return this.authenticationError ? 'warn' : 'primary';
            },
            enumerable: true,
            configurable: true
        });
        AuthComponent.prototype.ngOnInit = function () {
            var _this = this;
            if (common.isPlatformBrowser(this.platformId)) {
                this.onErrorSubscription = this.onError.subscribe(function () { return _this.authenticationError = true; });
            }
            this.min = this.min != null ? Math.max(this.min, this.config.passwordMinLength) : this.config.passwordMinLength;
            this.max = this.max != null ? Math.min(this.max, this.config.passwordMaxLength) : this.config.passwordMaxLength;
            this.goBackURL = this.chooseBackUrl();
            this.updateAuthSnackbarMessages();
            // auth form's initialization
            this._initSignInFormGroupBuilder();
            this._initSignUpFormGroupBuilder();
            this._initResetPasswordFormGroupBuilder();
        };
        AuthComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (this.passwordStrength) {
                this.passwordStrength.onStrengthChanged.subscribe(function (strength) {
                    _this.onStrengthChanged.emit(strength);
                });
            }
        };
        AuthComponent.prototype.ngOnChanges = function (changes) {
            if (changes.messageOnAuthSuccess || changes.messageOnAuthError) {
                this.updateAuthSnackbarMessages();
            }
            if (changes.min) {
                this.min = this.min != null ? Math.max(this.min, this.config.passwordMinLength) : this.config.passwordMinLength;
            }
            if (changes.max) {
                this.max = this.max != null ? Math.min(this.max, this.config.passwordMaxLength) : this.config.passwordMaxLength;
            }
            if (changes.goBackURL) {
                this.goBackURL = this.chooseBackUrl();
            }
        };
        AuthComponent.prototype.ngOnDestroy = function () {
            if (this.onErrorSubscription) {
                this.onErrorSubscription.unsubscribe();
            }
        };
        AuthComponent.prototype.onTabChange = function (event) {
            this.selectedTabChange.emit(event);
            this.tabIndex = event.index;
        };
        AuthComponent.prototype.signOut = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, , 2, 3]);
                            this.isLoading = true;
                            this.changeDetectorRef.markForCheck();
                            return [4 /*yield*/, this.authProcess.signOut()];
                        case 1:
                            _a.sent();
                            return [3 /*break*/, 3];
                        case 2:
                            this.isLoading = false;
                            this.tabIndex = 0;
                            this.changeDetectorRef.markForCheck();
                            return [7 /*endfinally*/];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        AuthComponent.prototype.signIn = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.signInFormGroup.valid) {
                                return [2 /*return*/];
                            }
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, , 3, 4]);
                            this.isLoading = true;
                            this.changeDetectorRef.markForCheck();
                            return [4 /*yield*/, this.authProcess.signInWith(this.authProviders.EmailAndPassword, {
                                    email: this.signInFormGroup.value.email,
                                    password: this.signInFormGroup.value.password
                                })];
                        case 2:
                            _a.sent();
                            return [3 /*break*/, 4];
                        case 3:
                            this.isLoading = false;
                            this.changeDetectorRef.markForCheck();
                            return [7 /*endfinally*/];
                        case 4: return [2 /*return*/];
                    }
                });
            });
        };
        AuthComponent.prototype.updateAuthSnackbarMessages = function () {
            this.authProcess.messageOnAuthSuccess = this.messageOnAuthSuccess;
            this.authProcess.messageOnAuthError = this.messageOnAuthError;
        };
        AuthComponent.prototype.createForgotPasswordTab = function () {
            this.passwordResetWished = true;
            this.tabIndex = 2;
            this.changeDetectorRef.markForCheck();
        };
        AuthComponent.prototype.processLegalSignUP = function (authProvider) {
            var _this = this;
            if (this.tosUrl || this.privacyPolicyUrl) {
                var params = {
                    tosUrl: this.tosUrl,
                    privacyPolicyUrl: this.privacyPolicyUrl,
                    authProvider: authProvider
                };
                this.dialogRef = this.dialog.open(LegalityDialogComponent, { data: params });
                this.dialogRef.afterClosed().subscribe(function (result) {
                    if (result && result.checked) {
                        _this._afterSignUpMiddleware(result.authProvider).then(function () { return _this.signUpFormGroup.reset(); });
                    }
                    _this.dialogRef = null;
                });
            }
            else {
                this._afterSignUpMiddleware(authProvider).then(function () { return _this.signUpFormGroup.reset(); });
            }
        };
        AuthComponent.prototype.signUp = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, , 2, 3]);
                            this.isLoading = true;
                            this.changeDetectorRef.markForCheck();
                            return [4 /*yield*/, this.authProcess.signUp(this.signUpFormGroup.value.name, {
                                    email: this.signUpFormGroup.value.email,
                                    password: this.signUpFormGroup.value.password
                                })];
                        case 1: return [2 /*return*/, _a.sent()];
                        case 2:
                            this.isLoading = false;
                            this.changeDetectorRef.markForCheck();
                            return [7 /*endfinally*/];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        AuthComponent.prototype.signUpAnonymously = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, , 2, 3]);
                            this.isLoading = true;
                            this.changeDetectorRef.markForCheck();
                            return [4 /*yield*/, this.authProcess.signInWith(this.authProvider.ANONYMOUS)];
                        case 1:
                            _a.sent();
                            return [3 /*break*/, 3];
                        case 2:
                            this.isLoading = false;
                            this.changeDetectorRef.markForCheck();
                            return [7 /*endfinally*/];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        AuthComponent.prototype.resetPassword = function () {
            var _this = this;
            this.authProcess.resetPassword(this.resetPasswordEmailFormControl.value)
                .then(function () {
                _this.passReset = true;
                // this.tabIndex = 2;
                _this.changeDetectorRef.markForCheck();
            });
        };
        AuthComponent.prototype.chooseBackUrl = function () {
            return this.activatedRoute.snapshot.queryParams.redirectUrl || this.goBackURL || '/';
        };
        AuthComponent.prototype._initSignInFormGroupBuilder = function () {
            this.signInFormGroup = new forms.FormGroup({});
            this.signInFormGroup.registerControl('email', this.signInEmailFormControl = new forms.FormControl('', [
                forms.Validators.required,
                forms.Validators.pattern(EMAIL_REGEX)
            ]));
            this.signInFormGroup.registerControl('password', this.sigInPasswordFormControl = new forms.FormControl('', [
                forms.Validators.required,
                forms.Validators.minLength(this.min),
                forms.Validators.maxLength(this.max)
            ]));
        };
        AuthComponent.prototype._initSignUpFormGroupBuilder = function () {
            this.signUpFormGroup = new forms.FormGroup({
                name: this.sigUpNameFormControl = new forms.FormControl('', [
                    forms.Validators.required,
                    forms.Validators.minLength(this.config.nameMinLength),
                    forms.Validators.maxLength(this.config.nameMaxLength)
                ]),
                email: this.sigUpEmailFormControl = new forms.FormControl('', [
                    forms.Validators.required,
                    forms.Validators.pattern(EMAIL_REGEX)
                ]),
                password: this.sigUpPasswordFormControl = new forms.FormControl('', [
                    forms.Validators.required,
                    forms.Validators.minLength(this.min),
                    forms.Validators.maxLength(this.max),
                ])
            });
        };
        AuthComponent.prototype._initResetPasswordFormGroupBuilder = function () {
            this.resetPasswordFormGroup = new forms.FormGroup({
                email: this.resetPasswordEmailFormControl = new forms.FormControl('', [
                    forms.Validators.required,
                    forms.Validators.pattern(EMAIL_REGEX)
                ])
            });
        };
        AuthComponent.prototype._afterSignUpMiddleware = function (authProvider) {
            if (authProvider === this.authProvider.ANONYMOUS) {
                return this.signUpAnonymously();
            }
            return this.signUp();
        };
        AuthComponent.ctorParameters = function () { return [
            { type: Object, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return NgxAuthFirebaseUIConfigToken; }),] }] },
            { type: auth.AngularFireAuth },
            { type: AuthProcessService },
            { type: dialog.MatDialog },
            { type: router.ActivatedRoute },
            { type: core.ChangeDetectorRef }
        ]; };
        __decorate([
            core.ViewChild(tabs.MatTabGroup, { static: false })
        ], AuthComponent.prototype, "matTabGroup", void 0);
        __decorate([
            core.ViewChild(passwordStrength.MatPasswordStrengthComponent, { static: false })
        ], AuthComponent.prototype, "passwordStrength", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "providers", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "providersTheme", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "appearance", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "tabIndex", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "registrationEnabled", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "resetPasswordEnabled", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "guestEnabled", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "tosUrl", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "privacyPolicyUrl", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "goBackURL", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "messageOnAuthSuccess", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "messageOnAuthError", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "messageOnEmailConfirmationSuccess", void 0);
        __decorate([
            core.Output()
        ], AuthComponent.prototype, "onSuccess", void 0);
        __decorate([
            core.Output()
        ], AuthComponent.prototype, "onError", void 0);
        __decorate([
            core.Output()
        ], AuthComponent.prototype, "selectedTabChange", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "enableLengthRule", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "enableLowerCaseLetterRule", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "enableUpperCaseLetterRule", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "enableDigitRule", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "enableSpecialCharRule", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "min", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "max", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "customValidator", void 0);
        __decorate([
            core.Output()
        ], AuthComponent.prototype, "onStrengthChanged", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "verifyEmailTemplate", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "verifyEmailTitleText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "verifyEmailConfirmationText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "verifyEmailGoBackText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "sendNewVerificationEmailText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "signOutText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "resetPasswordTabText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "resetPasswordInputText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "resetPasswordErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "resetPasswordErrorPatternText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "resetPasswordActionButtonText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "resetPasswordInstructionsText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "signInTabText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "signInCardTitleText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "loginButtonText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "forgotPasswordButtonText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "nameText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "nameErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "nameErrorMinLengthText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "nameErrorMaxLengthText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "emailText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "emailErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "emailErrorPatternText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "passwordText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "passwordErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "passwordErrorMinLengthText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "passwordErrorMaxLengthText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "registerTabText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "registerCardTitleText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "registerButtonText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "guestButtonText", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "emailConfirmationTitle", void 0);
        __decorate([
            core.Input()
        ], AuthComponent.prototype, "emailConfirmationText", void 0);
        AuthComponent = __decorate([
            core.Component({
                selector: 'ngx-auth-firebaseui',
                template: "<ng-container *ngIf=\"authProcess.user$ | async as user; else showForm\">\n\n  <!-- This component will be shown when:\n    - we just sent a verification mail (notably after sign up)\n    - we arrived from the guard after trying to access a protected route even though we are connected\n  -->\n  <div\n    *ngIf=\"(config.guardProtectedRoutesUntilEmailIsVerified && !user.emailVerified) || (authProcess.emailConfirmationSent && !user.emailVerified); else signedInUser\"\n    fxLayout=\"row\" fxLayoutAlign=\"center center\">\n    <ngx-auth-firebaseui-email-confirmation\n      (signOut)=\"signOut()\"\n      [email]=\"user.email\"\n      [goBackURL]=\"goBackURL\"\n      [messageOnEmailConfirmationSuccess]=\"messageOnEmailConfirmationSuccess\"\n      [sendNewVerificationEmailText]=\"sendNewVerificationEmailText\"\n      [signOutText]=\"signOutText\"\n      [template]=\"verifyEmailTemplate\"\n      [verifyEmailConfirmationText]=\"verifyEmailConfirmationText\"\n      [verifyEmailGoBackText]=\"verifyEmailGoBackText\"\n      [verifyEmailTitleText]=\"verifyEmailTitleText\">\n    </ngx-auth-firebaseui-email-confirmation>\n  </div>\n\n  <ng-template #signedInUser>\n    <div class=\"signed-in-container\" fxLayout=\"column\" fxLayoutAlign=\"center center\">\n      <img *ngIf=\"user?.photoURL; else noPhoto\" [src]=\"user?.photoURL\" class=\"account-circle\">\n      <ng-template #noPhoto>\n        <mat-icon class=\"account-circle\">account_circle</mat-icon>\n      </ng-template>\n      <div class=\"user-display-name mat-title\">{{ user?.displayName }}</div>\n      <div class=\"user-email mat-body-2\">{{ user?.email }}</div>\n      <div class=\"actions\">\n        <mat-progress-bar *ngIf=\"isLoading\" mode=\"indeterminate\"></mat-progress-bar>\n        <a *ngIf=\"verifyEmailGoBackText\" [routerLink]=\"goBackURL\" class=\"go-back-button action-button\" color=\"primary\"\n           mat-stroked-button>{{ verifyEmailGoBackText }}</a>\n        <button (click)=\"signOut()\" class=\"sign-out-button action-button\" color=\"warn\"\n                mat-stroked-button>{{ signOutText }}</button>\n      </div>\n    </div>\n  </ng-template>\n\n</ng-container>\n\n<ng-template #showForm>\n  <mat-tab-group (selectedTabChange)=\"onTabChange($event)\" [color]=\"color\" [selectedIndex]=\"tabIndex\">\n    <!--Sign in tab-->\n    <mat-tab [label]=\"signInTabText\">\n      <mat-card>\n        <mat-card-title>{{signInCardTitleText}}</mat-card-title>\n        <mat-card-content>\n          <form (ngSubmit)=\"signIn()\"\n                [@animateStagger]=\"{ value: '50' }\"\n                [formGroup]=\"signInFormGroup\">\n            <div fxLayout=\"column\" fxLayoutAlign=\"center\">\n              <mat-form-field [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n                              [appearance]=\"appearance\">\n                <mat-label>{{emailText}}</mat-label>\n                <input formControlName=\"email\"\n                       matInput\n                       required>\n                <mat-icon [color]=\"color\" matSuffix>email</mat-icon>\n                <mat-error *ngIf=\"signInEmailFormControl.hasError('required')\">\n                  {{emailErrorRequiredText}}\n                </mat-error>\n                <mat-error *ngIf=\"signInEmailFormControl.hasError('pattern')\">\n                  {{emailErrorPatternText}}\n                </mat-error>\n              </mat-form-field>\n\n              <mat-form-field [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n                              [appearance]=\"appearance\">\n                <mat-label>{{passwordText}}</mat-label>\n                <input [maxlength]=\"max\" [minlength]=\"min\" [type]=\"togglePass.type\" formControlName=\"password\" matInput\n                       required/>\n                <mat-pass-toggle-visibility #togglePass matSuffix></mat-pass-toggle-visibility>\n                <mat-icon [color]=\"color\" matSuffix>lock</mat-icon>\n                <mat-hint align=\"end\" aria-live=\"polite\"> {{ signInFormGroup.value.password.length }}\n                  / {{ max }} </mat-hint>\n                <mat-error *ngIf=\"sigInPasswordFormControl.hasError('required')\">\n                  {{passwordErrorRequiredText}}\n                </mat-error>\n                <mat-error *ngIf=\"sigInPasswordFormControl.hasError('minlength')\">\n                  {{ passwordErrorMinLengthText }}\n                </mat-error>\n                <mat-error *ngIf=\"sigInPasswordFormControl.hasError('maxlength')\">\n                  {{ passwordErrorMaxLengthText }}\n                </mat-error>\n              </mat-form-field>\n\n              <button [@animate]=\"{ value: '*', params: { x: '50px' } }\"\n                      [color]=\"color\"\n                      [disabled]=\"signInFormGroup.invalid\"\n                      class=\"space-top\"\n                      mat-raised-button\n                      style=\"margin-top: 20px\"\n                      type=\"submit\">\n                {{loginButtonText}}\n              </button>\n\n            </div>\n          </form>\n\n          <div fxLayoutAlign=\"center\">\n            <button (click)=\"createForgotPasswordTab()\"\n                    *ngIf=\"resetPasswordEnabled\"\n                    [@animate]=\"{ value: '*', params: { x: '-50px' } }\"\n                    [color]=\"color\"\n                    class=\"space-top\"\n                    mat-button>\n              {{forgotPasswordButtonText}}\n            </button>\n          </div>\n\n        </mat-card-content>\n        <mat-card-footer *ngIf=\"isLoading\">\n          <mat-progress-bar [@animate]=\"{ value: '*', params: { z: '50px', delay: '50ms', scale: '0.2' } }\"\n                            mode=\"indeterminate\"></mat-progress-bar>\n        </mat-card-footer>\n      </mat-card>\n    </mat-tab>\n\n    <!--tab register-->\n    <mat-tab *ngIf=\"registrationEnabled\" [label]=\"registerTabText\">\n      <mat-card>\n        <mat-card-title>{{registerCardTitleText}}</mat-card-title>\n        <mat-card-content fxLayout=\"column\" fxLayoutAlign=\"center\">\n          <form (ngSubmit)=\"signUpFormGroup.valid &&\n            processLegalSignUP(authProvider.EmailAndPassword)\"\n                [@animateStagger]=\"{ value: '50' }\" [formGroup]=\"signUpFormGroup\">\n            <div fxLayout=\"column\" fxLayoutAlign=\"center\">\n              <!--name-->\n              <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\"\n                              [appearance]=\"appearance\">\n                <!--labels will work only with @angular/material@6.2.0 -->\n                <mat-label>{{nameText}}</mat-label>\n                <input\n                  [formControl]=\"sigUpNameFormControl\"\n                  [maxlength]=\"config.nameMaxLength\"\n                  [minlength]=\"config.nameMinLength\"\n                  matInput\n                  required\n                />\n                <mat-icon [color]=\"color\" matSuffix>person</mat-icon>\n                <mat-hint align=\"end\" aria-live=\"polite\"> {{ signUpFormGroup.value.name?.length }}\n                  / {{ config.nameMaxLength }} </mat-hint>\n                <mat-error *ngIf=\"sigUpNameFormControl.hasError('required')\">\n                  {{nameErrorRequiredText}}\n                </mat-error>\n                <mat-error *ngIf=\"sigUpNameFormControl.hasError('minlength')\">\n                  {{nameErrorMinLengthText}}\n                </mat-error>\n                <mat-error *ngIf=\"sigUpNameFormControl.hasError('maxlength')\">\n                  {{nameErrorMaxLengthText}}\n                </mat-error>\n              </mat-form-field>\n\n              <!--email-->\n              <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\"\n                              [appearance]=\"appearance\">\n                <mat-label>{{emailText}}</mat-label>\n                <input [formControl]=\"sigUpEmailFormControl\"\n                       matInput\n                       required\n                       type=\"email\">\n                <mat-icon [color]=\"color\" matSuffix>email</mat-icon>\n                <mat-error *ngIf=\"sigUpEmailFormControl.hasError('required')\">\n                  {{emailErrorRequiredText}}\n                </mat-error>\n                <mat-error *ngIf=\"sigUpEmailFormControl.hasError('pattern')\">\n                  {{emailErrorPatternText}}\n                </mat-error>\n              </mat-form-field>\n\n              <!--password-->\n              <div fxLayout=\"column\">\n                <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\"\n                                [appearance]=\"appearance\">\n                  <mat-label>{{passwordText}}</mat-label>\n                  <input\n                    [formControl]=\"sigUpPasswordFormControl\"\n                    [maxlength]=\"max\"\n                    [minlength]=\"min\"\n                    [type]=\"toggle.type\"\n                    matInput\n                    name=\"password\"\n                    required\n                  />\n                  <mat-pass-toggle-visibility #toggle matSuffix></mat-pass-toggle-visibility>\n\n                  <mat-icon [color]=\"color\" matSuffix>lock</mat-icon>\n\n                  <mat-hint align=\"end\" aria-live=\"polite\">\n                    {{signUpFormGroup.value.password?.length}} / {{ max }}\n                  </mat-hint>\n\n                  <mat-error *ngIf=\"sigUpPasswordFormControl.hasError('required')\" class=\"cut-text\">\n                    {{passwordErrorRequiredText}}\n                  </mat-error>\n\n                  <mat-error *ngIf=\"sigUpPasswordFormControl.hasError('minlength')\" class=\"cut-text\">\n                    {{ passwordErrorMinLengthText }}\n                  </mat-error>\n                  <mat-error *ngIf=\"sigUpPasswordFormControl.hasError('maxlength')\" class=\"cut-text\">\n                    {{ passwordErrorMaxLengthText }}\n                  </mat-error>\n\n                </mat-form-field>\n\n                <mat-password-strength #passwordStrength\n                                       [customValidator]=\"customValidator\"\n                                       [enableDigitRule]=\"enableDigitRule\"\n                                       [enableLengthRule]=\"enableLengthRule\"\n                                       [enableLowerCaseLetterRule]=\"enableLowerCaseLetterRule\"\n                                       [enableSpecialCharRule]=\"enableSpecialCharRule\"\n                                       [enableUpperCaseLetterRule]=\"enableUpperCaseLetterRule\"\n                                       [externalError]=\"sigUpPasswordFormControl.dirty\"\n                                       [max]=\"max\"\n                                       [min]=\"min\"\n                                       [password]=\"signUpFormGroup.value.password\">\n                </mat-password-strength>\n\n              </div>\n\n              <button [@animate]=\"{ value: '*', params: { x: '100px' } }\"\n                      [color]=\"color\"\n                      [disabled]=\"signUpFormGroup.invalid\"\n                      mat-raised-button\n                      style=\"margin-top: 20px\"\n                      type=\"submit\">\n                {{registerButtonText}}\n              </button>\n\n            </div>\n          </form>\n\n          <button (click)=\"processLegalSignUP(authProvider.ANONYMOUS)\"\n                  *ngIf=\"guestEnabled\"\n                  [@animate]=\"{ value: '*', params: { x: '-100px' } }\"\n                  [color]=\"color\"\n                  mat-button\n                  style=\"margin-top: 20px\">\n            <mat-icon>fingerprint</mat-icon>\n            {{guestButtonText}}\n          </button>\n\n        </mat-card-content>\n\n        <mat-card-footer *ngIf=\"isLoading\">\n          <mat-progress-bar [@animate]=\"{ value: '*', params: { z: '50px', delay: '50ms', scale: '0.2' } }\"\n                            mode=\"indeterminate\"></mat-progress-bar>\n        </mat-card-footer>\n\n      </mat-card>\n    </mat-tab>\n\n    <!--Reset password tab-->\n    <mat-tab *ngIf=\"passwordResetWished\" class=\"reset-password-tab\">\n      <ng-template mat-tab-label>\n        <button (click)=\"passwordResetWished = false\" class=\"reset-password-tab__close-button\" mat-icon-button>\n          {{ resetPasswordTabText }}\n          <mat-icon>close</mat-icon>\n        </button>\n      </ng-template>\n      <form (ngSubmit)=\"resetPasswordFormGroup.valid && resetPassword()\"\n            [@animateStagger]=\"{ value: '50' }\"\n            [formGroup]=\"resetPasswordFormGroup\">\n        <mat-card class=\"reset-password-card\">\n          <mat-card-content>\n            <mat-form-field [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\" [appearance]=\"appearance\"\n                            class=\"full-width\">\n              <mat-label> {{ resetPasswordInputText }} </mat-label>\n              <input [title]=\"resetPasswordInputText\"\n                     formControlName=\"email\"\n                     matInput\n                     required>\n              <mat-icon [color]=\"color\" matSuffix>email</mat-icon>\n              <mat-error *ngIf=\"resetPasswordEmailFormControl.hasError('required')\">\n                {{resetPasswordErrorRequiredText}}\n              </mat-error>\n              <mat-error *ngIf=\"resetPasswordEmailFormControl.hasError('pattern')\">\n                {{resetPasswordErrorPatternText}}\n              </mat-error>\n            </mat-form-field>\n            <p *ngIf=\"passReset\">{{resetPasswordInstructionsText}}</p>\n          </mat-card-content>\n          <mat-card-actions fxLayoutAlign=\"center\">\n            <mat-progress-bar *ngIf=\"isLoading\" mode=\"indeterminate\"></mat-progress-bar>\n            <button [@animate]=\"{ value: '*', params: { x: '50px' } }\"\n                    [color]=\"color\"\n                    mat-raised-button\n                    type=\"submit\">\n              {{resetPasswordActionButtonText}}\n            </button>\n          </mat-card-actions>\n        </mat-card>\n      </form>\n    </mat-tab>\n\n  </mat-tab-group>\n  <mat-divider></mat-divider>\n  <ngx-auth-firebaseui-providers *ngIf=\"tabIndex !== 2\"\n                                 [providers]=\"providers\"\n                                 [theme]=\"providersTheme\"\n                                 [tosUrl]=\"tosUrl\"\n                                 [privacyPolicyUrl]=\"privacyPolicyUrl\">\n  </ngx-auth-firebaseui-providers>\n</ng-template>\n",
                animations: NgxAuthFirebaseuiAnimations,
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [".mat-card{margin:2rem}.space-top{margin-top:.5rem}.full-width{width:100%}.cut-text{text-overflow:ellipsis;overflow:hidden;white-space:nowrap}.signed-in-container .account-circle{font-size:12rem;width:12rem;height:12rem}.signed-in-container img.account-circle{-o-object-fit:cover;object-fit:cover;border-radius:50%}.signed-in-container .sign-out-button{margin-top:2rem}.signed-in-container .user-display-name{margin-top:1rem}.signed-in-container .user-email{margin-top:-1rem}.signed-in-container .actions{margin-top:2rem}.signed-in-container .actions .action-button,.signed-in-container .actions mat-progress-bar{width:100%}.signed-in-container .actions .action-button{margin-top:1rem}.reset-password-tab mat-progress-bar{margin-bottom:1rem}.reset-password-tab__close-button{width:100%;display:flex;justify-content:space-between;align-items:center}.reset-password-tab__close-button mat-icon{font-size:18px;position:relative;top:-1px}"]
            }),
            __param(0, core.Inject(core.PLATFORM_ID)),
            __param(1, core.Inject(core.forwardRef(function () { return NgxAuthFirebaseUIConfigToken; })))
        ], AuthComponent);
        return AuthComponent;
    }());

    var NgxAuthFirebaseuiAvatarComponent = /** @class */ (function () {
        function NgxAuthFirebaseuiAvatarComponent(afa, dialog) {
            this.afa = afa;
            this.dialog = dialog;
            this.layout = 'default';
            this.canLogout = true;
            this.canViewAccount = true;
            this.canDeleteAccount = true;
            this.canEditAccount = true;
            this.textProfile = 'Profile';
            this.textSignOut = 'Sign Out';
            // tslint:disable-next-line:no-output-on-prefix
            this.onSignOut = new core.EventEmitter();
        }
        NgxAuthFirebaseuiAvatarComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.user$ = this.afa.user;
            this.user$.subscribe(function (user) {
                _this.user = user;
                _this.displayNameInitials = user ? _this.getDisplayNameInitials(user.displayName) : null;
            });
        };
        NgxAuthFirebaseuiAvatarComponent.prototype.getDisplayNameInitials = function (displayName) {
            if (!displayName) {
                return null;
            }
            var initialsRegExp = displayName.match(/\b\w/g) || [];
            var initials = ((initialsRegExp.shift() || '') + (initialsRegExp.pop() || '')).toUpperCase();
            return initials;
        };
        NgxAuthFirebaseuiAvatarComponent.prototype.openProfile = function () {
            var dialogRef = this.dialog.open(UserComponent);
            var instance = dialogRef.componentInstance;
            instance.canDeleteAccount = this.canDeleteAccount;
            instance.canEditAccount = this.canEditAccount;
        };
        NgxAuthFirebaseuiAvatarComponent.prototype.signOut = function () {
            return __awaiter(this, void 0, void 0, function () {
                var e_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 2, , 3]);
                            return [4 /*yield*/, this.afa.signOut()];
                        case 1:
                            _a.sent();
                            // Sign-out successful.
                            this.onSignOut.emit();
                            return [3 /*break*/, 3];
                        case 2:
                            e_1 = _a.sent();
                            // An error happened.
                            console.error('An error happened while signing out!', e_1);
                            return [3 /*break*/, 3];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        NgxAuthFirebaseuiAvatarComponent.ctorParameters = function () { return [
            { type: auth.AngularFireAuth },
            { type: dialog.MatDialog }
        ]; };
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "layout", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "canLogout", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "links", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "canViewAccount", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "canDeleteAccount", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "canEditAccount", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "textProfile", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "textSignOut", void 0);
        __decorate([
            core.Output()
        ], NgxAuthFirebaseuiAvatarComponent.prototype, "onSignOut", void 0);
        NgxAuthFirebaseuiAvatarComponent = __decorate([
            core.Component({
                selector: 'ngx-auth-firebaseui-avatar',
                template: "<button *ngIf=\"user\"\n        class=\"ngx-auth-firebaseui-avatar-button\"\n        [matMenuTriggerFor]=\"posXMenu\"\n        [matTooltip]=\"user?.displayName\"\n        [style.background-image]=\"'url(' + user?.photoURL + ')'\"\n        aria-label=\"Open x-positioned menu\"\n        mat-mini-fab\n        style=\"background-size: cover\">\n  <span *ngIf=\"!user?.photoURL\">{{displayNameInitials || ''}}</span>\n</button>\n\n<mat-menu #posXMenu=\"matMenu\" class=\"before ngx-auth-firebaseui-avatar-menu\" xPosition=\"before\" >\n  <div fxLayout=\"row\" fxLayout.xs=\"column\" style=\"padding-left: 10px; padding-right: 10px\" [ngStyle]=\"{ 'padding-top.px': layout === 'default' ? 0 : 10 }\">\n    <button [style.background-image]=\"'url(' + user?.photoURL + ')'\"\n            mat-fab\n            style=\"background-size: cover\"\n            *ngIf=\"layout === 'default'\">\n      <span *ngIf=\"!user?.photoURL\">{{displayNameInitials || ''}}</span>\n    </button>\n    <div fxLayout=\"column\" style=\"padding-left: 10px; padding-right: 10px\">\n      <strong mat-card-title>{{user?.displayName}}</strong>\n      <em mat-card-subtitle style=\"font-style: italic\">{{user?.email}}</em>\n    </div>\n  </div>\n\n  <div fxFlex=\"100\" fxLayout=\"column\" [ngStyle]=\"{ 'padding-bottom.px': layout === 'default' ? 0 : 10 } \">\n    <div *ngFor=\"let menuItem of links\" class=\"links-menu\">\n      <button (click)=\"menuItem?.callback()\" mat-menu-item>\n        <mat-icon>{{menuItem?.icon}}</mat-icon>\n        {{menuItem?.text}}</button>\n    </div>\n    <button *ngIf=\"canViewAccount\" (click)=\"openProfile()\" color=\"primary\" fxLayoutAlign=\"center\" mat-raised-button>{{ textProfile }}\n    </button>\n    <button (click)=\"signOut()\" *ngIf=\"canLogout\" color=\"warn\" fxLayoutAlign=\"center\" mat-raised-button>{{ textSignOut }}\n    </button>\n  </div>\n</mat-menu>\n",
                styles: [".mat-raised-button{margin:.2rem 1rem}.links-menu{text-align:center}"]
            })
        ], NgxAuthFirebaseuiAvatarComponent);
        return NgxAuthFirebaseuiAvatarComponent;
    }());

    var NgxAuthFirebaseuiLoginComponent = /** @class */ (function () {
        function NgxAuthFirebaseuiLoginComponent(
        // tslint:disable-next-line:ban-types
        platformId, authProcess, formBuilder) {
            this.platformId = platformId;
            this.authProcess = authProcess;
            this.formBuilder = formBuilder;
            this.providers = exports.AuthProvider.ALL; //  google, facebook, twitter, github as array or all as one single string
            this.registrationEnabled = true;
            this.resetPasswordEnabled = true;
            // i18n
            this.titleText = 'LOGIN TO YOUR ACCOUNT';
            this.rememberMeText = 'Remember Me';
            this.loginButtonText = 'LOGIN';
            this.orLabelText = 'OR';
            this.forgotPasswordText = 'Forgot Password?';
            this.dontHaveAnAccountText = 'Don\'t have an account?';
            this.createAccountButtonText = 'Create an account';
            // i18n email
            this.emailText = 'Email';
            this.emailErrorRequiredText = 'Email is required';
            this.emailErrorPatternText = 'Please enter a valid email address';
            // i18n password
            this.passwordText = 'Password';
            this.passwordErrorRequiredText = 'Password is required';
            // tslint:disable-next-line:no-output-on-prefix
            this.onCreateAccountRequested = new core.EventEmitter();
            // tslint:disable-next-line:no-output-on-prefix
            this.onResetPasswordRequested = new core.EventEmitter();
            this.authProviders = exports.AuthProvider;
            this.authenticationError = false;
            this.onSuccess = authProcess.onSuccessEmitter;
            this.onError = authProcess.onErrorEmitter;
        }
        Object.defineProperty(NgxAuthFirebaseuiLoginComponent.prototype, "color", {
            get: function () {
                return this.authenticationError ? 'warn' : 'primary';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxAuthFirebaseuiLoginComponent.prototype, "colorAccent", {
            get: function () {
                return this.authenticationError ? 'warn' : 'accent';
            },
            enumerable: true,
            configurable: true
        });
        NgxAuthFirebaseuiLoginComponent.prototype.ngOnInit = function () {
            var _this = this;
            if (common.isPlatformBrowser(this.platformId)) {
                this.onErrorSubscription = this.onError.subscribe(function () { return _this.authenticationError = true; });
            }
            this.updateAuthSnackbarMessages();
            this.loginForm = this.formBuilder.group({
                email: ['', [forms.Validators.required, forms.Validators.email]],
                password: ['', forms.Validators.required]
            });
        };
        NgxAuthFirebaseuiLoginComponent.prototype.updateAuthSnackbarMessages = function () {
            this.authProcess.messageOnAuthSuccess = this.messageOnAuthSuccess;
            this.authProcess.messageOnAuthError = this.messageOnAuthError;
        };
        NgxAuthFirebaseuiLoginComponent.prototype.login = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.authProcess.signInWith(this.authProviders.EmailAndPassword, {
                                email: this.loginForm.controls.email.value,
                                password: this.loginForm.controls.password.value
                            })];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            });
        };
        NgxAuthFirebaseuiLoginComponent.ctorParameters = function () { return [
            { type: Object, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] },
            { type: AuthProcessService },
            { type: forms.FormBuilder }
        ]; };
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "logoUrl", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "providers", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "appearance", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "registrationEnabled", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "resetPasswordEnabled", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "messageOnAuthSuccess", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "messageOnAuthError", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "titleText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "rememberMeText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "loginButtonText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "orLabelText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "forgotPasswordText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "dontHaveAnAccountText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "createAccountButtonText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "emailText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "emailErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "emailErrorPatternText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "passwordText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "passwordErrorRequiredText", void 0);
        __decorate([
            core.Output()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "onSuccess", void 0);
        __decorate([
            core.Output()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "onError", void 0);
        __decorate([
            core.Output()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "onCreateAccountRequested", void 0);
        __decorate([
            core.Output()
        ], NgxAuthFirebaseuiLoginComponent.prototype, "onResetPasswordRequested", void 0);
        NgxAuthFirebaseuiLoginComponent = __decorate([
            core.Component({
                selector: 'ngx-auth-firebaseui-login',
                template: "<div fxLayout=\"column\" id=\"login\">\n\n  <div fxLayout=\"column\" fxLayoutAlign=\"center center\" id=\"login-form-wrapper\">\n\n    <div [@animateStagger]=\"{ value: '50' }\" id=\"login-form\">\n\n      <div *ngIf=\"logoUrl\" class=\"logo\">\n        <img [@animate]=\"{ value: '*', params: { x: '50px' } }\" [src]=\"logoUrl\" alt=\"logo\">\n      </div>\n\n      <div [@animate]=\"{ value: '*', params: { x: '-50px' } }\" class=\"title\">{{titleText}}</div>\n\n      <form [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\" [formGroup]=\"loginForm\" name=\"loginForm\"\n            novalidate>\n\n        <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\" [appearance]=\"appearance\">\n          <input [placeholder]=\"emailText\" formControlName=\"email\" matInput>\n          <mat-icon [color]=\"color\" matSuffix>email</mat-icon>\n          <mat-error *ngIf=\"loginForm.get('email')?.hasError('required')\">\n            {{emailErrorRequiredText}}\n          </mat-error>\n          <mat-error\n            *ngIf=\"!loginForm.get('email')?.hasError('required') &&\n                                loginForm.get('email')?.hasError('email')\">\n            {{emailErrorPatternText}}\n          </mat-error>\n        </mat-form-field>\n\n        <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\" [appearance]=\"appearance\">\n          <input [placeholder]=\"passwordText\" formControlName=\"password\" matInput type=\"password\">\n          <mat-icon [color]=\"color\" matSuffix>lock</mat-icon>\n          <mat-error>\n            {{passwordErrorRequiredText}}\n          </mat-error>\n        </mat-form-field>\n\n        <div [@animate]=\"{ value: '*', params: { x: '50px' } }\"\n             class=\"remember-forgot-password\" fxLayout=\"row\"\n             fxLayout.xs=\"column\"\n             fxLayoutAlign=\"space-between center\">\n          <!--          <mat-checkbox class=\"remember-me\" aria-label=\"Remember Me\">-->\n          <!--            {{rememberMeText}}-->\n          <!--          </mat-checkbox>-->\n\n          <button (click)=\"onResetPasswordRequested.emit()\"\n                  *ngIf=\"resetPasswordEnabled\"\n                  [@animate]=\"{ value: '*', params: { x: '-50px' } }\"\n                  [color]=\"color\"\n                  class=\"forgot-password\"\n                  mat-button\n                  type=\"button\">\n            {{forgotPasswordText}}\n          </button>\n        </div>\n\n        <button (click)=\"login()\"\n                [color]=\"colorAccent\"\n                [disabled]=\"loginForm.invalid\"\n                aria-label=\"LOG IN\"\n                class=\"submit-button\"\n                id=\"loginButton\"\n                mat-raised-button>\n          {{loginButtonText}}\n        </button>\n\n      </form>\n\n      <div *ngIf=\"providers.length > 0\"\n           [@animate]=\"{ value: '*', params: { z: '50px', delay: '50ms', scale: '0.2' } }\"\n           class=\"separator\">\n        <span class=\"text\">{{orLabelText}}</span>\n      </div>\n\n      <ngx-auth-firebaseui-providers [providers]=\"providers\"\n                                     fxLayoutAlign=\"center center\"\n                                     layout=\"column\"\n                                     theme=\"raised\"></ngx-auth-firebaseui-providers>\n\n      <div *ngIf=\"registrationEnabled\"\n           [@animateStagger]=\"{ value: '100' }\"\n           class=\"register\"\n           fxLayout=\"column\" fxLayoutAlign=\"center center\">\n        <span [@animate]=\"{ value: '*', params: { x: '100px' } }\" class=\"text\">\n          {{dontHaveAnAccountText}}\n        </span>\n        <button (click)=\"onCreateAccountRequested.emit()\"\n                [@animate]=\"{ value: '*', params: { x: '-100px' } }\"\n                [color]=\"color\"\n                id=\"createAccountButton\"\n                mat-button\n                type=\"button\">{{createAccountButtonText}}</button>\n      </div>\n    </div>\n  </div>\n</div>\n",
                encapsulation: core.ViewEncapsulation.None,
                animations: NgxAuthFirebaseuiAnimations,
                styles: ["ngx-auth-firebaseui-login #login-form-wrapper{flex:1 0 auto;padding:32px}ngx-auth-firebaseui-login #login-form-wrapper #login-form{width:384px;max-width:384px;padding:32px;text-align:center}ngx-auth-firebaseui-login #login-form-wrapper #login-form .logo{width:150px;height:150px;margin:32px auto}ngx-auth-firebaseui-login #login-form-wrapper #login-form .title{font-size:20px;margin:16px 0 32px}ngx-auth-firebaseui-login #login-form-wrapper #login-form form{width:100%;text-align:left}ngx-auth-firebaseui-login #login-form-wrapper #login-form form mat-form-field{width:100%}ngx-auth-firebaseui-login #login-form-wrapper #login-form form mat-checkbox{margin:0}ngx-auth-firebaseui-login #login-form-wrapper #login-form form .remember-forgot-password{font-size:13px;margin-top:8px}ngx-auth-firebaseui-login #login-form-wrapper #login-form form .remember-forgot-password .remember-me{margin-bottom:16px}ngx-auth-firebaseui-login #login-form-wrapper #login-form form .remember-forgot-password .forgot-password{font-size:13px;font-weight:500;margin-bottom:16px}ngx-auth-firebaseui-login #login-form-wrapper #login-form form .submit-button{width:220px;margin:16px auto;display:block}@media screen and (max-width:599px){ngx-auth-firebaseui-login #login-form-wrapper{padding:16px}ngx-auth-firebaseui-login #login-form-wrapper #login-form{padding:24px;width:100%}ngx-auth-firebaseui-login #login-form-wrapper #login-form form .submit-button{width:90%}ngx-auth-firebaseui-login #login-form-wrapper #login-form button{width:80%}}ngx-auth-firebaseui-login #login-form-wrapper #login-form .register{margin:32px auto 24px;font-weight:500}ngx-auth-firebaseui-login #login-form-wrapper #login-form .register .text{margin-right:8px}ngx-auth-firebaseui-login #login-form-wrapper #login-form .separator{font-size:15px;font-weight:600;margin:24px auto;position:relative;overflow:hidden;width:100px}ngx-auth-firebaseui-login #login-form-wrapper #login-form .separator .text{display:inline-flex;position:relative;padding:0 8px;z-index:9999}ngx-auth-firebaseui-login #login-form-wrapper #login-form .separator .text:after,ngx-auth-firebaseui-login #login-form-wrapper #login-form .separator .text:before{content:\"\";display:block;width:30px;position:absolute;top:10px;border-top:1px solid}ngx-auth-firebaseui-login #login-form-wrapper #login-form .separator .text:before{right:100%}ngx-auth-firebaseui-login #login-form-wrapper #login-form .separator .text:after{left:100%}ngx-auth-firebaseui-login #login-form-wrapper #login-form button.facebook-raised,ngx-auth-firebaseui-login #login-form-wrapper #login-form button.github-raised,ngx-auth-firebaseui-login #login-form-wrapper #login-form button.google-raised,ngx-auth-firebaseui-login #login-form-wrapper #login-form button.microsoft-raised,ngx-auth-firebaseui-login #login-form-wrapper #login-form button.twitter-raised,ngx-auth-firebaseui-login #login-form-wrapper #login-form button.yahoo-raised{width:192px;text-transform:none;color:#fff;font-size:13px;margin-bottom:8px}"]
            }),
            __param(0, core.Inject(core.PLATFORM_ID))
        ], NgxAuthFirebaseuiLoginComponent);
        return NgxAuthFirebaseuiLoginComponent;
    }());

    var confirmPasswordValidator = function (control) {
        if (!control.parent || !control) {
            return null;
        }
        var password = control.parent.get('password');
        var passwordConfirm = control.parent.get('passwordConfirm');
        if (!password || !passwordConfirm) {
            return null;
        }
        if (passwordConfirm.value === '') {
            return null;
        }
        if (password.value === passwordConfirm.value) {
            return null;
        }
        return { passwordsNotMatching: true };
    };
    var NgxAuthFirebaseuiRegisterComponent = /** @class */ (function () {
        // tslint:disable-next-line:ban-types
        function NgxAuthFirebaseuiRegisterComponent(platformId, formBuilder, authProcess) {
            // Configure the layout
            this.platformId = platformId;
            this.formBuilder = formBuilder;
            this.authProcess = authProcess;
            // i18n common
            this.titleText = 'CREATE AN ACCOUNT';
            this.readAndAcceptText = 'I read and accept the';
            this.termsAndConditionsText = 'terms and conditions';
            this.privacyPolicyText = 'privacy policy';
            this.createAccountButtonText = 'CREATE AN ACCOUNT';
            this.alreadyHaveAccountText = 'Already have an account?';
            this.loginButtonText = 'LOGIN';
            // i18n name
            this.nameText = 'Name';
            this.nameErrorRequiredText = 'Name is required';
            // i18n email
            this.emailText = 'Email';
            this.emailErrorRequiredText = 'Email is required';
            this.emailErrorPatternText = 'Please enter a valid email address';
            // i18n password
            this.passwordText = 'Password';
            this.passwordErrorRequiredText = 'Password is required';
            this.passwordConfirmationText = 'Password Confirmation';
            this.passwordConfirmationErrorRequiredText = 'Password confirmation is required';
            this.passwordErrorMatchText = 'Password must match';
            // tslint:disable-next-line:no-output-on-prefix
            this.onLoginRequested = new core.EventEmitter();
            this.authenticationError = false;
            // Set the private defaults
            this.unsubscribeAll = new rxjs.Subject();
            this.onSuccess = authProcess.onSuccessEmitter;
            this.onError = authProcess.onErrorEmitter;
        }
        Object.defineProperty(NgxAuthFirebaseuiRegisterComponent.prototype, "color", {
            get: function () {
                return this.authenticationError ? 'warn' : 'primary';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxAuthFirebaseuiRegisterComponent.prototype, "colorAccent", {
            get: function () {
                return this.authenticationError ? 'warn' : 'accent';
            },
            enumerable: true,
            configurable: true
        });
        NgxAuthFirebaseuiRegisterComponent.prototype.ngOnInit = function () {
            var _this = this;
            if (common.isPlatformBrowser(this.platformId)) {
                this.onErrorSubscription = this.onError.subscribe(function () { return _this.authenticationError = true; });
            }
            this.registerForm = this.formBuilder.group({
                name: ['', forms.Validators.required],
                email: ['', [forms.Validators.required, forms.Validators.email]],
                password: ['', forms.Validators.required],
                passwordConfirm: ['', [forms.Validators.required, confirmPasswordValidator]],
                tos: [''],
                privacyPolicy: ['']
            });
            // If tos or privacy policy url set, ensure that the two form items are required
            if (this.tosUrl) {
                this.registerForm.controls.tos.setValidators(forms.Validators.requiredTrue);
            }
            if (this.privacyPolicyUrl) {
                this.registerForm.controls.privacyPolicy.setValidators(forms.Validators.requiredTrue);
            }
            // Update the validity of the 'passwordConfirm' field
            // when the 'password' field changes
            this.registerForm
                .controls
                .password
                .valueChanges.pipe(operators.takeUntil(this.unsubscribeAll))
                .subscribe(function () {
                _this.registerForm.controls.passwordConfirm.updateValueAndValidity();
            });
        };
        /**
         * On destroy
         */
        NgxAuthFirebaseuiRegisterComponent.prototype.ngOnDestroy = function () {
            // Unsubscribe from all subscriptions
            this.unsubscribeAll.next();
            this.unsubscribeAll.complete();
        };
        NgxAuthFirebaseuiRegisterComponent.prototype.createAccount = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.authProcess.signUp(this.registerForm.controls.name.value, {
                                email: this.registerForm.controls.email.value,
                                password: this.registerForm.controls.password.value
                            })];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            });
        };
        NgxAuthFirebaseuiRegisterComponent.ctorParameters = function () { return [
            { type: Object, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] },
            { type: forms.FormBuilder },
            { type: AuthProcessService }
        ]; };
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "logoUrl", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "appearance", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "tosUrl", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "privacyPolicyUrl", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "titleText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "readAndAcceptText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "termsAndConditionsText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "privacyPolicyText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "createAccountButtonText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "alreadyHaveAccountText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "loginButtonText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "nameText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "nameErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "emailText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "emailErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "emailErrorPatternText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "passwordText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "passwordErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "passwordConfirmationText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "passwordConfirmationErrorRequiredText", void 0);
        __decorate([
            core.Input()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "passwordErrorMatchText", void 0);
        __decorate([
            core.Output()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "onSuccess", void 0);
        __decorate([
            core.Output()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "onError", void 0);
        __decorate([
            core.Output()
        ], NgxAuthFirebaseuiRegisterComponent.prototype, "onLoginRequested", void 0);
        NgxAuthFirebaseuiRegisterComponent = __decorate([
            core.Component({
                selector: 'ngx-auth-firebaseui-register',
                template: "<div fxLayout=\"column\" id=\"register\">\n\n  <div fxLayout=\"column\" fxLayoutAlign=\"center center\" id=\"register-form-wrapper\">\n\n    <div [@animateStagger]=\"{ value: '50' }\" id=\"register-form\">\n\n      <div *ngIf=\"logoUrl\" class=\"logo\">\n        <img [@animate]=\"{ value: '*', params: { x: '50px' } }\" [src]=\"logoUrl\" alt=\"logo\">\n      </div>\n\n      <div [@animate]=\"{ value: '*', params: { x: '-50px' } }\" class=\"title\">{{titleText}}</div>\n\n      <form [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\" [formGroup]=\"registerForm\" name=\"registerForm\"\n            novalidate>\n\n        <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\" [appearance]=\"appearance\">\n          <input [placeholder]=\"nameText\" formControlName=\"name\" matInput/>\n          <mat-icon [color]=\"color\" matSuffix>person</mat-icon>\n          <mat-error>\n            {{nameErrorRequiredText}}\n          </mat-error>\n        </mat-form-field>\n\n        <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\" [appearance]=\"appearance\">\n          <input [placeholder]=\"emailText\" formControlName=\"email\" matInput/>\n          <mat-icon [color]=\"color\" matSuffix>email</mat-icon>\n          <mat-error *ngIf=\"registerForm.get('email')?.hasError('required')\">\n            {{emailErrorRequiredText}}\n          </mat-error>\n          <mat-error *ngIf=\"registerForm.get('email')?.hasError('email')\">\n            {{emailErrorPatternText}}\n          </mat-error>\n        </mat-form-field>\n\n        <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\" [appearance]=\"appearance\">\n          <input [placeholder]=\"passwordText\" formControlName=\"password\" matInput type=\"password\"/>\n          <mat-icon [color]=\"color\" matSuffix>lock</mat-icon>\n          <mat-error>\n            {{passwordErrorRequiredText}}\n          </mat-error>\n        </mat-form-field>\n\n        <mat-form-field [@animate]=\"{ value: '*', params: { x: '50px' } }\" [appearance]=\"appearance\">\n          <input [placeholder]=\"passwordConfirmationText\" formControlName=\"passwordConfirm\" matInput type=\"password\"/>\n          <mat-icon [color]=\"color\" matSuffix>lock</mat-icon>\n          <mat-error *ngIf=\"registerForm.get('passwordConfirm')?.hasError('required')\">\n            {{passwordConfirmationErrorRequiredText}}\n          </mat-error>\n          <mat-error\n            *ngIf=\"\n              !registerForm.get('passwordConfirm')?.hasError('required') &&\n              registerForm.get('passwordConfirm')?.hasError('passwordsNotMatching')\n            \">\n            {{passwordErrorMatchText}}\n          </mat-error>\n        </mat-form-field>\n\n        <div *ngIf=\"this.tosUrl\">\n          <mat-checkbox aria-label=\"{{readAndAcceptText}}\" formControlName=\"tos\" required>\n            <span>{{readAndAcceptText}}</span>\n            <a target=\"_blank\" [href]=\"this.tosUrl\">\n                {{termsAndConditionsText}}\n            </a>\n          </mat-checkbox>\n        </div>\n\n        <div *ngIf=\"this.privacyPolicyUrl\">\n          <mat-checkbox aria-label=\"{{readAndAcceptText}}\" formControlName=\"privacyPolicy\" required>\n            <span>{{readAndAcceptText}}</span>\n            <a target=\"_blank\" [href]=\"this.privacyPolicyUrl\">\n                {{privacyPolicyText}}\n            </a>\n          </mat-checkbox>\n        </div>\n\n        <button (click)=\"createAccount()\"\n                [color]=\"colorAccent\"\n                [disabled]=\"registerForm.invalid\"\n                aria-label=\"CREATE AN ACCOUNT\"\n                class=\"submit-button\"\n                id=\"createAccountButton\"\n                mat-raised-button>\n          {{createAccountButtonText}}\n        </button>\n      </form>\n\n      <div [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\" class=\"register\" fxLayout=\"column\"\n           fxLayoutAlign=\"center center\">\n        <span [@animate]=\"{ value: '*', params: { x: '100px' } }\" class=\"text\">\n          {{alreadyHaveAccountText}}\n        </span>\n        <button (click)=\"onLoginRequested.emit()\"\n                [@animate]=\"{ value: '*', params: { x: '-100px' } }\"\n                [color]=\"colorAccent\"\n                id=\"loginButton\"\n                mat-button\n                type=\"button\">\n          {{loginButtonText}}\n        </button>\n      </div>\n\n    </div>\n  </div>\n</div>\n",
                encapsulation: core.ViewEncapsulation.None,
                animations: NgxAuthFirebaseuiAnimations,
                styles: ["ngx-auth-firebaseui-register #register{width:100%;background-size:cover}ngx-auth-firebaseui-register #register #register-form-wrapper{flex:1 0 auto;padding:32px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form{width:384px;max-width:384px;padding:32px;text-align:center}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .logo{width:128px;margin:32px auto}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .title{font-size:20px;margin:16px 0 32px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form form{width:100%;text-align:left}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form form mat-form-field{width:100%}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form form mat-checkbox{margin:0}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form form .terms{margin:16px 0 32px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form form .terms a{font-size:16px;margin-left:4px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form form .submit-button{width:220px;margin:16px auto;display:block}@media screen and (max-width:599px){ngx-auth-firebaseui-register #register #register-form-wrapper{padding:16px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form{padding:24px;width:100%}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form form .submit-button{width:90%}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form button{width:80%}}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .register{margin:32px auto 24px;font-weight:500}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .register .text{margin-right:8px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .separator{font-size:15px;font-weight:600;margin:24px auto;position:relative;overflow:hidden;width:100px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .separator .text{display:inline-flex;position:relative;padding:0 8px;z-index:9999}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .separator .text:after,ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .separator .text:before{content:\"\";display:block;width:30px;position:absolute;top:10px;border-top:1px solid}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .separator .text:before{right:100%}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form .separator .text:after{left:100%}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form button.facebook,ngx-auth-firebaseui-register #register #register-form-wrapper #register-form button.google{width:192px;text-transform:none;color:#fff;font-size:13px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form button.google{background-color:#d73d32;margin-bottom:8px}ngx-auth-firebaseui-register #register #register-form-wrapper #register-form button.facebook{background-color:#3f5c9a}ngx-auth-firebaseui-register ::ng-deep .mat-checkbox-label{display:flex;flex-wrap:wrap}"]
            }),
            __param(0, core.Inject(core.PLATFORM_ID))
        ], NgxAuthFirebaseuiRegisterComponent);
        return NgxAuthFirebaseuiRegisterComponent;
    }());

    var UserComponent = /** @class */ (function () {
        function UserComponent(auth, authProcess, fireStoreService, config) {
            this.auth = auth;
            this.authProcess = authProcess;
            this.fireStoreService = fireStoreService;
            this.config = config;
            this.canLogout = true;
            this.canEditAccount = true;
            this.canDeleteAccount = true;
            // tslint:disable-next-line:no-output-on-prefix
            this.onSignOut = new core.EventEmitter();
            // tslint:disable-next-line:no-output-on-prefix
            this.onAccountEdited = new core.EventEmitter();
            // tslint:disable-next-line:no-output-on-prefix
            this.onAccountDeleted = new core.EventEmitter();
        }
        UserComponent.prototype.changeEditMode = function () {
            var _this = this;
            if (this.editMode) {
                this.reset();
                this.editMode = false;
            }
            else {
                this.initUpdateFormGroup().subscribe(function (updateFormGroup) {
                    _this.updateFormGroup = updateFormGroup;
                    _this.editMode = true;
                });
            }
        };
        UserComponent.prototype.reset = function () {
            this.updateFormGroup.reset();
            this.updateFormGroup.disable();
            this.updateFormGroup = null;
        };
        UserComponent.prototype.save = function () {
            return __awaiter(this, void 0, void 0, function () {
                var user, snackBarMsg, error_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.updateFormGroup.dirty) return [3 /*break*/, 12];
                            this.editMode = false;
                            user = this.authProcess.user;
                            snackBarMsg = [];
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 10, , 11]);
                            if (!this.updateNameFormControl.dirty) return [3 /*break*/, 3];
                            return [4 /*yield*/, user.updateProfile({ displayName: this.updateNameFormControl.value })];
                        case 2:
                            _a.sent();
                            snackBarMsg.push("your name has been updated to " + user.displayName);
                            _a.label = 3;
                        case 3:
                            if (!this.updateEmailFormControl.dirty) return [3 /*break*/, 5];
                            return [4 /*yield*/, user.updateEmail(this.updateEmailFormControl.value)];
                        case 4:
                            _a.sent();
                            snackBarMsg.push("your email has been updated to " + user.email);
                            _a.label = 5;
                        case 5:
                            if (!this.updatePhoneNumberFormControl.dirty) return [3 /*break*/, 7];
                            return [4 /*yield*/, user.updatePhoneNumber(this.updatePhoneNumberFormControl.value)];
                        case 6:
                            _a.sent();
                            console.log('phone number = ', this.updatePhoneNumberFormControl.value);
                            snackBarMsg.push("your phone number has been updated to " + user.phoneNumber);
                            _a.label = 7;
                        case 7:
                            if (!this.config.enableFirestoreSync) return [3 /*break*/, 9];
                            return [4 /*yield*/, this.fireStoreService.updateUserData(this.authProcess.parseUserInfo(user))];
                        case 8:
                            _a.sent();
                            _a.label = 9;
                        case 9: return [3 /*break*/, 11];
                        case 10:
                            error_1 = _a.sent();
                            this.authProcess.showToast(error_1 && error_1.message ? error_1.message : error_1);
                            console.error(error_1);
                            return [3 /*break*/, 11];
                        case 11:
                            if (snackBarMsg.length > 0) {
                                this.authProcess.showToast(snackBarMsg.join('\\n'));
                            }
                            this.updateFormGroup.reset();
                            _a.label = 12;
                        case 12: return [2 /*return*/];
                    }
                });
            });
        };
        UserComponent.prototype.signOut = function () {
            var _this = this;
            this.auth.signOut()
                .then(function () { return _this.onSignOut.emit(); })
                .catch(function (e) { return console.error('An error happened while signing out!', e); });
        };
        /**
         * Delete the account of the current firebase ngx-auth-firebaseui-user
         *
         * On Success, emit the <onAccountDeleted> event and toast a msg!#
         * Otherwise, log the and toast and error msg!
         *
         */
        UserComponent.prototype.deleteAccount = function () {
            return __awaiter(this, void 0, void 0, function () {
                var user, error_2;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 3, , 4]);
                            user = this.authProcess.user;
                            // await this.authProcess.deleteAccount();
                            return [4 /*yield*/, this.authProcess.user.delete()];
                        case 1:
                            // await this.authProcess.deleteAccount();
                            _a.sent();
                            // if (this.config.enableFirestoreSync) {
                            return [4 /*yield*/, this.fireStoreService.deleteUserData(user.uid)];
                        case 2:
                            // if (this.config.enableFirestoreSync) {
                            _a.sent();
                            // }
                            this.onAccountDeleted.emit();
                            this.editMode = false;
                            console.log('Your account has been successfully deleted!');
                            this.authProcess.showToast('Your account has been successfully deleted!');
                            return [3 /*break*/, 4];
                        case 3:
                            error_2 = _a.sent();
                            console.log('Error while delete user account', error_2);
                            this.authProcess.showToast("Error occurred while deleting your account: " + error_2.message);
                            return [3 /*break*/, 4];
                        case 4: return [2 /*return*/];
                    }
                });
            });
        };
        UserComponent.prototype.initUpdateFormGroup = function () {
            var _this = this;
            return this.authProcess.user$.pipe(operators.take(1), operators.map(function (currentUser) {
                var updateFormGroup = new forms.FormGroup({
                    name: _this.updateNameFormControl = new forms.FormControl({ value: currentUser.displayName, disabled: _this.editMode }, [
                        forms.Validators.required,
                        forms.Validators.minLength(_this.config.nameMinLength),
                        forms.Validators.maxLength(_this.config.nameMaxLength)
                    ]),
                    email: _this.updateEmailFormControl = new forms.FormControl({ value: currentUser.email, disabled: _this.editMode }, [
                        forms.Validators.required,
                        forms.Validators.pattern(EMAIL_REGEX)
                    ]),
                    phoneNumber: _this.updatePhoneNumberFormControl = new forms.FormControl({ value: currentUser.phoneNumber, disabled: _this.editMode }, [forms.Validators.pattern(PHONE_NUMBER_REGEX)])
                });
                updateFormGroup.enable();
                return updateFormGroup;
            }));
        };
        UserComponent.ctorParameters = function () { return [
            { type: auth.AngularFireAuth },
            { type: AuthProcessService },
            { type: FirestoreSyncService },
            { type: undefined, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return NgxAuthFirebaseUIConfigToken; }),] }] }
        ]; };
        __decorate([
            core.Input()
        ], UserComponent.prototype, "editMode", void 0);
        __decorate([
            core.Input()
        ], UserComponent.prototype, "canLogout", void 0);
        __decorate([
            core.Input()
        ], UserComponent.prototype, "canEditAccount", void 0);
        __decorate([
            core.Input()
        ], UserComponent.prototype, "canDeleteAccount", void 0);
        __decorate([
            core.Input()
        ], UserComponent.prototype, "appearance", void 0);
        __decorate([
            core.Output()
        ], UserComponent.prototype, "onSignOut", void 0);
        __decorate([
            core.Output()
        ], UserComponent.prototype, "onAccountEdited", void 0);
        __decorate([
            core.Output()
        ], UserComponent.prototype, "onAccountDeleted", void 0);
        UserComponent = __decorate([
            core.Component({
                selector: 'ngx-auth-firebaseui-user',
                template: "<div *ngIf=\"auth.authState| async; then authenticated else none\">\n\n</div>\n\n<ng-template #authenticated>\n  <mat-card *ngIf=\"auth.user | async as user\">\n    <!--<form [formGroup]=\"updateFormGroup\" >-->\n    <!--card header-->\n    <mat-card-header fxLayout=\"column\" fxLayoutAlign=\"center center\">\n\n      <img *ngIf=\"authProcess?.getUserPhotoUrl() | async as photoUrl\" [src]=\"photoUrl\" mat-card-avatar>\n\n      <div *ngIf=\"user.emailVerified; then emailVerified else emailNotVerified\"></div>\n      <ng-template #emailVerified>\n        <mat-icon color=\"primary\"\n                  matTooltip=\"email is verified\"\n                  matTooltipPosition=\"after\">\n          verified_user\n        </mat-icon>\n      </ng-template>\n      <ng-template #emailNotVerified>\n        <mat-icon color=\"warn\"\n                  matTooltip=\"email is not verified\"\n                  matTooltipPosition=\"after\">\n          warning\n        </mat-icon>\n      </ng-template>\n\n    </mat-card-header>\n\n    <!--card content-->\n    <mat-card-content *ngIf=\"editMode; then edit else readonly\">\n    </mat-card-content>\n\n    <ng-template #edit>\n      <form (submit)=\"save()\" [formGroup]=\"updateFormGroup\">\n\n        <mat-card-content fxLayout=\"column\" fxLayoutAlign=\"center center\">\n          <div fxLayoutAlign=\"center\">\n            <button (click)=\"changeEditMode()\" class=\"edit-button\" color=\"warn\"\n                    mat-raised-button>\n              cancel\n            </button>\n          </div>\n\n          <!--name-->\n          <mat-form-field [appearance]=\"appearance\" class=\"full-width\">\n            <mat-label>Name</mat-label>\n            <input [formControl]=\"updateNameFormControl\"\n                   matInput\n                   placeholder=\"Name\">\n            <mat-icon matSuffix>person</mat-icon>\n            <mat-hint align=\"end\" aria-live=\"polite\"> {{ updateNameFormControl.value?.length }}\n              / {{ config.nameMaxLength }} </mat-hint>\n            <mat-error *ngIf=\"updateNameFormControl.hasError('required')\">\n              Name is required\n            </mat-error>\n          </mat-form-field>\n\n          <!--email-->\n          <mat-form-field [appearance]=\"appearance\" class=\"full-width\">\n            <mat-label>E-mail</mat-label>\n            <input [formControl]=\"updateEmailFormControl\"\n                   matInput\n                   placeholder=\"E-mail\">\n            <mat-icon matSuffix>email</mat-icon>\n            <mat-error *ngIf=\"updateEmailFormControl.hasError('required')\">\n              E-mail is required {{updateEmailFormControl.value}}\n            </mat-error>\n            <mat-error *ngIf=\"updateEmailFormControl.hasError('pattern')\">\n              Please enter a valid e-mail address {{updateEmailFormControl.value}}\n            </mat-error>\n          </mat-form-field>\n\n          <!--phone number-->\n          <mat-form-field *ngIf=\"false\" [appearance]=\"appearance\" class=\"full-width\">\n            <mat-label>Phone number</mat-label>\n            <input [formControl]=\"updatePhoneNumberFormControl\"\n                   matInput\n                   placeholder=\"Phone number\"\n                   type=\"tel\">\n            <mat-icon matSuffix>phone</mat-icon>\n            <mat-hint align=\"end\" aria-live=\"polite\">\n              The phone number is international. Therefore, it should start with a + sign or 00,\n              followed by the country code, - and national number e.g: +49-12345678 or 0041-1234567890\n\n              NOTE : the phone number must be a valid phone credential !!\n            </mat-hint>\n            <mat-error *ngIf=\"updatePhoneNumberFormControl.hasError('pattern')\">\n              Please enter a valid phone number\n            </mat-error>\n          </mat-form-field>\n\n        </mat-card-content>\n\n        <mat-card-actions fxLayout=\"column\">\n          <button color=\"primary\"\n                  mat-button\n                  type=\"submit\">\n            Save changes\n          </button>\n        </mat-card-actions>\n      </form>\n    </ng-template>\n\n    <ng-template #readonly>\n      <div fxLayoutAlign=\"center\">\n        <button *ngIf=\"canEditAccount\" (click)=\"changeEditMode()\" class=\"edit-button\" color=\"primary\"\n                mat-raised-button>\n          edit\n        </button>\n      </div>\n\n      <!--name-->\n      <mat-form-field [appearance]=\"appearance\" class=\"full-width\">\n        <mat-label>Name</mat-label>\n        <input [disabled]=\"!editMode\"\n               [value]=\"user.displayName\"\n               matInput\n               placeholder=\"Name\">\n        <mat-icon color=\"primary\" matSuffix>person</mat-icon>\n      </mat-form-field>\n\n      <!--email-->\n      <mat-form-field [appearance]=\"appearance\" class=\"full-width\">\n        <mat-label>E-mail</mat-label>\n        <input [disabled]=\"!editMode\"\n               [value]=\"user.email\" matInput\n               placeholder=\"E-mail\">\n        <mat-icon color=\"primary\" matSuffix>email</mat-icon>\n      </mat-form-field>\n\n      <!--phone number-->\n      <mat-form-field *ngIf=\"false\" [appearance]=\"appearance\" class=\"full-width\">\n        <mat-label>Phone number</mat-label>\n        <input [disabled]=\"!editMode\"\n               [value]=\"user.phoneNumber\"\n               matInput\n               placeholder=\"Phone number\">\n        <mat-icon color=\"primary\" matSuffix>phone</mat-icon>\n      </mat-form-field>\n\n      <mat-card-actions fxLayout=\"column\">\n        <button (click)=\"signOut()\" *ngIf=\"canLogout\" color=\"primary\" mat-button>Sign out</button>\n        <button (click)=\"deleteAccount()\" *ngIf=\"canDeleteAccount\" color=\"warn\" mat-button>Delete account</button>\n      </mat-card-actions>\n\n    </ng-template>\n\n  </mat-card>\n\n</ng-template>\n\n\n<ng-template #none>\n  <mat-card class=\"none-card\" fxLayout=\"row\" fxLayoutAlign=\"center center\">\n    <mat-card-content fxLayout=\"row\" fxLayoutAlign=\"center center\">\n      <mat-icon color=\"accent\">warning</mat-icon>\n      <span>You are not logged in!</span>\n    </mat-card-content>\n  </mat-card>\n</ng-template>\n",
                styles: [".edit-button{margin:1rem}.full-width{width:100%}.cut-text{text-overflow:ellipsis;overflow:hidden;white-space:nowrap}.none-card{min-height:430px}.none-card span{font-size:24px;text-align:center;color:rgba(0,0,0,.54)}"]
            }),
            __param(3, core.Inject(core.forwardRef(function () { return NgxAuthFirebaseUIConfigToken; })))
        ], UserComponent);
        return UserComponent;
    }());


    (function (Theme) {
        Theme["DEFAULT"] = "default";
        Theme["CLASSIC"] = "classic";
        Theme["STROKED"] = "stroked";
        Theme["FAB"] = "fab";
        Theme["MINI_FAB"] = "mini-fab";
        Theme["RAISED"] = "raised";
    })(exports.Theme || (exports.Theme = {}));

    (function (Layout) {
        Layout["ROW"] = "row";
        Layout["COLUMN"] = "column";
    })(exports.Layout || (exports.Layout = {}));
    var AuthProvidersComponent = /** @class */ (function () {
        function AuthProvidersComponent(authProcess, dialog) {
            this.authProcess = authProcess;
            this.dialog = dialog;
            this.layout = exports.Layout.ROW;
            this.providers = exports.AuthProvider.ALL; //  google, facebook, twitter, github, microsoft, yahoo
            this.themes = exports.Theme;
            this.authProvider = exports.AuthProvider;
            this.onSuccess = authProcess.onSuccessEmitter;
            this.onError = authProcess.onErrorEmitter;
        }
        AuthProvidersComponent.prototype.processLegalSignUP = function (authProvider) {
            var _this = this;
            if (this.tosUrl || this.privacyPolicyUrl) {
                var params = {
                    tosUrl: this.tosUrl,
                    privacyPolicyUrl: this.privacyPolicyUrl,
                    authProvider: authProvider
                };
                this.dialogRef = this.dialog.open(LegalityDialogComponent, { data: params });
                this.dialogRef.afterClosed().subscribe(function (result) {
                    if (result && result.checked) {
                        // this._afterSignUpMiddleware(result.authProvider).then(() => this.signUpFormGroup.reset());
                        _this.authProcess.signInWith(authProvider);
                    }
                    _this.dialogRef = null;
                });
            }
            else {
                // this._afterSignUpMiddleware(authProvider).then(() => this.signUpFormGroup.reset());
                this.authProcess.signInWith(authProvider);
            }
        };
        AuthProvidersComponent.ctorParameters = function () { return [
            { type: AuthProcessService },
            { type: dialog.MatDialog }
        ]; };
        __decorate([
            core.Input()
        ], AuthProvidersComponent.prototype, "theme", void 0);
        __decorate([
            core.Input()
        ], AuthProvidersComponent.prototype, "layout", void 0);
        __decorate([
            core.Input()
        ], AuthProvidersComponent.prototype, "providers", void 0);
        __decorate([
            core.Output()
        ], AuthProvidersComponent.prototype, "onSuccess", void 0);
        __decorate([
            core.Output()
        ], AuthProvidersComponent.prototype, "onError", void 0);
        __decorate([
            core.Input()
        ], AuthProvidersComponent.prototype, "tosUrl", void 0);
        __decorate([
            core.Input()
        ], AuthProvidersComponent.prototype, "privacyPolicyUrl", void 0);
        AuthProvidersComponent = __decorate([
            core.Component({
                selector: 'ngx-auth-firebaseui-providers',
                template: "<div [@animateStagger]=\"{ value: '50' }\" [ngSwitch]=\"theme\">\n\n  <!--default icon buttons-->\n  <div *ngSwitchDefault\n       [fxLayoutAlign]=\"layout == 'row' ? 'space-around center' : 'stretch'\"\n       [fxLayout]=\"layout\"\n       fxLayout.xs=\"column\">\n    <button (click)=\"processLegalSignUP(authProvider.Google)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Google)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            mat-button>\n      <mat-icon svgIcon=\"google-colored\"></mat-icon>\n      Google\n    </button>\n\n    <button (click)=\"processLegalSignUP(authProvider.Apple)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Apple)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"apple-filled\"\n            mat-button>\n      <mat-icon svgIcon=\"apple\"></mat-icon>\n      Apple\n    </button>\n\n    <button (click)=\"processLegalSignUP(authProvider.Facebook)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Facebook)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"facebook-filled\"\n            mat-button>\n      <mat-icon svgIcon=\"facebook\"></mat-icon>\n      Facebook\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Twitter)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Twitter)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"twitter-filled\"\n            mat-button>\n      <mat-icon svgIcon=\"twitter\"></mat-icon>\n      Twitter\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Github)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Github)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            mat-button>\n      <mat-icon svgIcon=\"github\"></mat-icon>\n      GitHub\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Microsoft)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Microsoft)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            mat-button>\n      <mat-icon svgIcon=\"microsoft\"></mat-icon>\n      Microsoft\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Yahoo)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Yahoo)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            mat-button>\n      <mat-icon svgIcon=\"yahoo\"></mat-icon>\n      Yahoo\n    </button>\n  </div>\n\n  <!--classic-->\n  <div *ngSwitchCase=\"themes.CLASSIC\"\n       [fxLayoutAlign]=\"layout == 'row' ? 'space-around center' : 'stretch'\"\n       [fxLayout]=\"layout\"\n       class=\"buttons-classic\"\n       fxLayout.xs=\"column\">\n    <button (click)=\"processLegalSignUP(authProvider.Google)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Google)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"google-classic\"\n            mat-button>\n      Google\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Apple)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Apple)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"apple-classic\"\n            mat-button>\n      Apple\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Facebook)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Facebook)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"facebook-classic\"\n            mat-button>\n      Facebook\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Twitter)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Twitter)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"twitter-classic\"\n            mat-button>\n      Twitter\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Github)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Github)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"github-classic\"\n            mat-button>\n      GitHub\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Microsoft)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Microsoft)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"microsoft-classic\"\n            mat-button>\n      Microsoft\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Yahoo)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Yahoo)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"yahoo-classic\"\n            mat-button>\n      Yahoo\n    </button>\n  </div>\n\n  <!--stroked-->\n  <div *ngSwitchCase=\"themes.STROKED\"\n       [fxLayoutAlign]=\"layout == 'row' ? 'space-around center' : 'stretch'\"\n       [fxLayout]=\"layout\"\n       class=\"buttons-classic\"\n       fxLayout.xs=\"column\">\n    <button (click)=\"processLegalSignUP(authProvider.Google)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Google)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"google-classic\"\n            mat-stroked-button>\n      Google\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Apple)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Apple)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"apple-classic\"\n            mat-stroked-button>\n      Apple\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Facebook)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Facebook)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"facebook-classic\"\n            mat-stroked-button>\n      Facebook\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Twitter)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Twitter)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"twitter-classic\"\n            mat-stroked-button>\n      Twitter\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Github)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Github)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"github-classic\"\n            mat-stroked-button>\n      GitHub\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Microsoft)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Microsoft)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"microsoft-classic\"\n            mat-stroked-button>\n      Microsoft\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Yahoo)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Yahoo)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"yahoo-classic\"\n            mat-stroked-button>\n      Yahoo\n    </button>\n  </div>\n\n  <!--raised-->\n  <div *ngSwitchCase=\"themes.RAISED\"\n       [fxLayoutAlign]=\"layout == 'row' ? 'space-around center' : 'stretch'\"\n       [fxLayout]=\"layout\"\n       class=\"buttons-raised\"\n       fxLayout.xs=\"column\">\n    <button (click)=\"processLegalSignUP(authProvider.Google)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Google)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"google-raised\"\n            mat-raised-button>\n      Google\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Apple)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Apple)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"apple-raised\"\n            mat-raised-button>\n      Apple\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Facebook)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Facebook)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"facebook-raised\"\n            mat-raised-button>\n      Facebook\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Twitter)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Twitter)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"twitter-raised\"\n            mat-raised-button>\n      Twitter\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Github)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Github)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"github-raised\"\n            mat-raised-button>\n      GitHub\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Microsoft)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Microsoft)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"microsoft-raised\"\n            mat-raised-button>\n      Microsoft\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Yahoo)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Yahoo)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            [ngClass.xs]=\"{'space-full-xs':true}\"\n            class=\"yahoo-raised\"\n            mat-raised-button>\n      Yahoo\n    </button>\n  </div>\n\n  <!--fab-->\n  <div *ngSwitchCase=\"themes.FAB\"\n       [fxLayoutAlign]=\"layout == 'row' ? 'space-around center' : 'stretch'\"\n       [fxLayout]=\"layout\"\n       class=\"buttons-raised\">\n    <button (click)=\"processLegalSignUP(authProvider.Google)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Google)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"google-raised\"\n            mat-fab>\n      <mat-icon svgIcon=\"google\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Apple)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Apple)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"apple-raised\"\n            mat-fab>\n      <mat-icon svgIcon=\"apple\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Facebook)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Facebook)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"facebook-raised\"\n            mat-fab>\n      <mat-icon svgIcon=\"facebook\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Twitter)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Twitter)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"twitter-raised\"\n            mat-fab>\n      <mat-icon svgIcon=\"twitter\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Github)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Github)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"github-raised\"\n            mat-fab>\n      <mat-icon svgIcon=\"github\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Microsoft)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Microsoft)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"microsoft\"\n            mat-fab>\n      <mat-icon svgIcon=\"microsoft\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Yahoo)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Yahoo)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"yahoo-raised\"\n            mat-fab>\n      <mat-icon svgIcon=\"yahoo\"></mat-icon>\n    </button>\n  </div>\n\n  <!--mini-fab-->\n  <div *ngSwitchCase=\"themes.MINI_FAB\"\n       [fxLayoutAlign]=\"layout == 'row' ? 'space-around center' : 'stretch'\"\n       [fxLayout]=\"layout\"\n       class=\"buttons-raised\"\n       fxLayoutAlign.xs=\"center center\">\n    <button (click)=\"processLegalSignUP(authProvider.Google)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Google)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"google-raised\"\n            fxFlexAlign=\"center\"\n            mat-mini-fab>\n      <mat-icon svgIcon=\"google\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Apple)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Apple)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"apple-raised\"\n            mat-mini-fab>\n      <mat-icon svgIcon=\"apple\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Facebook)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Facebook)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"facebook-raised\"\n            mat-mini-fab>\n      <mat-icon svgIcon=\"facebook\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Twitter)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Twitter)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"twitter-raised\"\n            mat-mini-fab>\n      <mat-icon class=\"icon-white\" svgIcon=\"twitter\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Github)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Github)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"github-raised\"\n            mat-mini-fab>\n      <mat-icon svgIcon=\"github\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Microsoft)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Microsoft)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"microsoft\"\n            mat-mini-fab>\n      <mat-icon svgIcon=\"microsoft\"></mat-icon>\n    </button>\n    <button (click)=\"processLegalSignUP(authProvider.Yahoo)\"\n            *ngIf=\"providers === authProvider.ALL || providers.includes(authProvider.Yahoo)\"\n            [@animate]=\"{value:'*',params:{duration:'300ms',y:'100px'}}\"\n            class=\"yahoo-raised\"\n            mat-mini-fab>\n      <mat-icon svgIcon=\"yahoo\"></mat-icon>\n    </button>\n  </div>\n</div>\n",
                animations: NgxAuthFirebaseuiAnimations,
                styles: [":host{display:block}.space-full-xs{width:100%;margin:.4rem}.apple-filled mat-icon svg path{fill:#000}.facebook-filled mat-icon{fill:#385899}.twitter-filled mat-icon{fill:#1da1f2}.buttons-raised button{color:#fff!important}.buttons-raised .google-raised{background-color:#db4437}.buttons-raised .apple-raised{background-color:#000}.buttons-raised .facebook-raised{background-color:#385899}.buttons-raised .twitter-raised{background-color:#1da1f2}.buttons-raised .github-raised{background-color:#000}.buttons-raised .microsoft-raised{background-color:#0078d4}.buttons-raised .yahoo-raised{background-color:#720e9e}.buttons-raised .phone-raised{background-color:#02bd7e}.buttons-classic button.google-classic{color:#db4437!important}.buttons-classic button.apple-classic{color:#000!important}.buttons-classic .facebook-classic{color:#385899!important}.buttons-classic .twitter-classic{color:#1da1f2!important}.buttons-classic .github-classic{color:#000!important}.buttons-classic .microsoft-classic{color:#0078d4!important}.buttons-classic .yahoo-classic{color:#720e9e!important}.buttons-classic .phone-classic{color:#02bd7e}.icon-white{color:#fff}.icon-white mat-icon{fill:#fff}button.microsoft{background:#f8f9fa}"]
            })
        ], AuthProvidersComponent);
        return AuthProvidersComponent;
    }());

    // import * as firebase from 'firebase';
    var defaultAuthFirebaseUIConfig = {
        // authMethod: 'redirect',
        // authProviders: [new GoogleAuthProvider(), new FacebookAuthProvider(), new TwitterAuthProvider(), new GithubAuthProvider()],
        enableFirestoreSync: true,
        toastMessageOnAuthSuccess: true,
        toastMessageOnAuthError: true,
        authGuardFallbackURL: '/',
        authGuardLoggedInURL: '/',
        // Password length min/max in forms independently of each componenet min/max.
        // `min/max` input parameters in components should be within this range.
        passwordMaxLength: 60,
        passwordMinLength: 8,
        // Same as password but for the name
        nameMaxLength: 50,
        nameMinLength: 2,
        // If set, sign-in/up form is not available until email has been verified.
        // Plus protected routes are still protected even though user is connected.
        guardProtectedRoutesUntilEmailIsVerified: true,
        // Default to email verification on
        enableEmailVerification: true
    };
    // Merge default config with user provided config.
    function ngxAuthFirebaseUIConfigFactory(userProvidedConfig) {
        return Object.assign({}, defaultAuthFirebaseUIConfig, userProvidedConfig);
    }

    var LoggedInGuard = /** @class */ (function () {
        function LoggedInGuard(config, router, authProcess) {
            this.config = config;
            this.router = router;
            this.authProcess = authProcess;
        }
        LoggedInGuard.prototype.canActivate = function (route, state) {
            var _this = this;
            return this.authProcess.afa.user.pipe(operators.map(function (user) {
                if (user) {
                    if (_this.config.guardProtectedRoutesUntilEmailIsVerified && !user.emailVerified && !user.isAnonymous) {
                        if (_this.config.authGuardFallbackURL) {
                            _this.router.navigate(["" + _this.config.authGuardFallbackURL], { queryParams: { redirectUrl: state.url } });
                        }
                        return false;
                    }
                    else {
                        return true;
                    }
                }
                else {
                    if (_this.config.authGuardFallbackURL) {
                        _this.router.navigate(["/" + _this.config.authGuardFallbackURL], { queryParams: { redirectUrl: state.url } });
                    }
                    return false;
                }
            }));
        };
        LoggedInGuard.ctorParameters = function () { return [
            { type: undefined, decorators: [{ type: core.Inject, args: [NgxAuthFirebaseUIConfigToken,] }] },
            { type: router.Router },
            { type: AuthProcessService }
        ]; };
        LoggedInGuard.ɵprov = core.ɵɵdefineInjectable({ factory: function LoggedInGuard_Factory() { return new LoggedInGuard(core.ɵɵinject(NgxAuthFirebaseUIConfigToken), core.ɵɵinject(router.Router), core.ɵɵinject(AuthProcessService)); }, token: LoggedInGuard, providedIn: "root" });
        LoggedInGuard = __decorate([
            core.Injectable({
                providedIn: 'root'
            }),
            __param(0, core.Inject(NgxAuthFirebaseUIConfigToken))
        ], LoggedInGuard);
        return LoggedInGuard;
    }());

    var NgxAuthFirebaseUIModule = /** @class */ (function () {
        function NgxAuthFirebaseUIModule(iconRegistry, sanitizer, auth) {
            this.iconRegistry = iconRegistry;
            this.sanitizer = sanitizer;
            auth.listenToUserEvents();
            this.registerProviderIcons();
        }
        NgxAuthFirebaseUIModule_1 = NgxAuthFirebaseUIModule;
        NgxAuthFirebaseUIModule.forRoot = function (configFactory, appNameFactory, config) {
            if (appNameFactory === void 0) { appNameFactory = function () { return undefined; }; }
            if (config === void 0) { config = {}; }
            return {
                ngModule: NgxAuthFirebaseUIModule_1,
                providers: [
                    {
                        provide: fire.FIREBASE_OPTIONS,
                        useValue: configFactory
                    },
                    {
                        provide: fire.FIREBASE_APP_NAME,
                        useFactory: appNameFactory
                    },
                    { provide: UserProvidedConfigToken, useValue: config },
                    {
                        provide: NgxAuthFirebaseUIConfigToken,
                        useFactory: ngxAuthFirebaseUIConfigFactory,
                        deps: [UserProvidedConfigToken]
                    },
                    AuthProcessService,
                    FirestoreSyncService,
                    LoggedInGuard
                ]
            };
        };
        NgxAuthFirebaseUIModule.prototype.registerProviderIcons = function () {
            this.iconRegistry
                .addSvgIcon('google', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/mdi/google.svg'))
                .addSvgIcon('apple', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/mdi/apple.svg'))
                .addSvgIcon('google-colored', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/google.svg'))
                .addSvgIcon('facebook', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/mdi/facebook.svg'))
                .addSvgIcon('twitter', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/mdi/twitter.svg'))
                .addSvgIcon('github', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/mdi/github-circle.svg'))
                .addSvgIcon('microsoft', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/mdi/microsoft.svg'))
                .addSvgIcon('yahoo', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/mdi/yahoo.svg'))
                .addSvgIcon('phone', this.sanitizer.bypassSecurityTrustResourceUrl('/assets/phone.svg'));
        };
        var NgxAuthFirebaseUIModule_1;
        NgxAuthFirebaseUIModule.ctorParameters = function () { return [
            { type: icon.MatIconRegistry },
            { type: platformBrowser.DomSanitizer },
            { type: AuthProcessService }
        ]; };
        NgxAuthFirebaseUIModule = NgxAuthFirebaseUIModule_1 = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    // HTTP
                    router.RouterModule,
                    http.HttpClientModule,
                    // FLEX_LAYOUT
                    flexLayout.FlexLayoutModule,
                    // FORMS
                    forms.FormsModule,
                    forms.ReactiveFormsModule,
                    // MATERIAL2
                    tabs.MatTabsModule,
                    card.MatCardModule,
                    input.MatInputModule,
                    button.MatButtonModule,
                    icon.MatIconModule,
                    snackBar.MatSnackBarModule,
                    divider.MatDividerModule,
                    chips.MatChipsModule,
                    tooltip.MatTooltipModule,
                    dialog.MatDialogModule,
                    checkbox.MatCheckboxModule,
                    progressSpinner.MatProgressSpinnerModule,
                    progressBar.MatProgressBarModule,
                    dialog.MatDialogModule,
                    menu.MatMenuModule,
                    // ANGULAR MATERIAL EXTENSIONS
                    passwordStrength.MatPasswordStrengthModule,
                    // ANGULARFIRE2
                    auth.AngularFireAuthModule,
                    firestore.AngularFirestoreModule,
                ],
                exports: [
                    AuthComponent,
                    UserComponent,
                    NgxAuthFirebaseuiAvatarComponent,
                    AuthProvidersComponent,
                    EmailConfirmationComponent,
                    // LoggedInGuard,
                    auth.AngularFireAuthModule,
                    firestore.AngularFirestoreModule,
                    NgxAuthFirebaseuiLoginComponent,
                    NgxAuthFirebaseuiRegisterComponent
                ],
                declarations: [
                    AuthComponent,
                    UserComponent,
                    NgxAuthFirebaseuiAvatarComponent,
                    AuthProvidersComponent,
                    EmailConfirmationComponent,
                    LegalityDialogComponent,
                    NgxAuthFirebaseuiLoginComponent,
                    NgxAuthFirebaseuiRegisterComponent
                ],
                entryComponents: [
                    UserComponent,
                    LegalityDialogComponent
                ]
            })
        ], NgxAuthFirebaseUIModule);
        return NgxAuthFirebaseUIModule;
    }());

    exports.AuthComponent = AuthComponent;
    exports.AuthProcessService = AuthProcessService;
    exports.AuthProvidersComponent = AuthProvidersComponent;
    exports.EMAIL_REGEX = EMAIL_REGEX;
    exports.EmailConfirmationComponent = EmailConfirmationComponent;
    exports.FirestoreSyncService = FirestoreSyncService;
    exports.LegalityDialogComponent = LegalityDialogComponent;
    exports.LoggedInGuard = LoggedInGuard;
    exports.NgxAuthFirebaseUIConfigToken = NgxAuthFirebaseUIConfigToken;
    exports.NgxAuthFirebaseUIModule = NgxAuthFirebaseUIModule;
    exports.NgxAuthFirebaseuiAvatarComponent = NgxAuthFirebaseuiAvatarComponent;
    exports.NgxAuthFirebaseuiLoginComponent = NgxAuthFirebaseuiLoginComponent;
    exports.NgxAuthFirebaseuiRegisterComponent = NgxAuthFirebaseuiRegisterComponent;
    exports.PHONE_NUMBER_REGEX = PHONE_NUMBER_REGEX;
    exports.UserComponent = UserComponent;
    exports.UserProvidedConfigToken = UserProvidedConfigToken;
    exports.appleAuthProvider = appleAuthProvider;
    exports.collections = collections;
    exports.confirmPasswordValidator = confirmPasswordValidator;
    exports.defaultAuthFirebaseUIConfig = defaultAuthFirebaseUIConfig;
    exports.facebookAuthProvider = facebookAuthProvider;
    exports.githubAuthProvider = githubAuthProvider;
    exports.googleAuthProvider = googleAuthProvider;
    exports.microsoftAuthProvider = microsoftAuthProvider;
    exports.ngxAuthFirebaseUIConfigFactory = ngxAuthFirebaseUIConfigFactory;
    exports.twitterAuthProvider = twitterAuthProvider;
    exports.yahooAuthProvider = yahooAuthProvider;
    exports.ɵa = defaultAuthFirebaseUIConfig;
    exports.ɵb = ngxAuthFirebaseUIConfigFactory;
    exports.ɵd = NgxAuthFirebaseuiAnimations;
    exports.ɵe = EmailConfirmationComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
